// glimmer-benchmarks/lib/benchmarks/components.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var bench_1 = require("../bench");
var NestedEmberishCurleyTaglessComponentsScenario = (function (_super) {
    __extends(NestedEmberishCurleyTaglessComponentsScenario, _super);
    function NestedEmberishCurleyTaglessComponentsScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = "nested emberish curley tagless component";
        _this.description = "10 outer components, each with 10 inner components";
        return _this;
    }
    NestedEmberishCurleyTaglessComponentsScenario.prototype.start = function () {
        this.glimmerEnv.registerStaticTaglessComponent("parent-component", null, "<h1>{{parent.name}}</h1>\n        <ul>\n          {{#each parent.children key=\"@index\" as |child|}}\n            {{child-component child=child parent=parent}}\n          {{/each}}\n        </ul>");
        this.glimmerEnv.registerStaticTaglessComponent("child-component", null, "<li>{{child.name}} (from {{parent.name}})</li>");
        _super.prototype.start.call(this);
    };
    NestedEmberishCurleyTaglessComponentsScenario.prototype.template = function () {
        return "{{#each parents key=\"@index\" as |parent|}}\n              {{parent-component parent=parent}}\n            {{/each}}";
    };
    NestedEmberishCurleyTaglessComponentsScenario.prototype.renderContext = function () {
        var context = {
            parents: []
        };
        for (var i = 0; i < 10; i++) {
            var parent = { name: "Parent " + i, children: [] };
            context.parents.push(parent);
            for (var j = 0; j < 10; j++) {
                parent.children.push({ name: "Child " + j });
            }
        }
        return context;
    };
    NestedEmberishCurleyTaglessComponentsScenario.prototype.test = function (render) {
        var element = render();
        var has10Parents = element.getElementsByTagName('h1').length === 10;
        var hasParentText = element.getElementsByTagName('h1')[0].textContent === 'Parent 0';
        var has100Children = element.getElementsByTagName('li').length === 100;
        var hasChildText = element.getElementsByTagName('li')[0].textContent === 'Child 0 (from Parent 0)';
        if (!has10Parents || !hasParentText || !has100Children || !hasChildText) {
            throw new Error("Invalid render: " + element.outerHTML);
        }
    };
    return NestedEmberishCurleyTaglessComponentsScenario;
}(bench_1.TemplateBenchmarkScenario));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = [
    NestedEmberishCurleyTaglessComponentsScenario
];
