// glimmer-benchmarks/lib/benchmarks/if.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var bench_1 = require("../bench");
var IfScenario = (function (_super) {
    __extends(IfScenario, _super);
    function IfScenario() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IfScenario.prototype.start = function () {
        this.glimmerEnv.registerHelper('inline-if', function (_a) {
            var cond = _a[0], truthy = _a[1], falsy = _a[2];
            return cond ? truthy : falsy;
        });
        _super.prototype.start.call(this);
    };
    IfScenario.prototype.renderContext = function () {
        return {
            truthy: true,
            falsy: false,
            first: "Kris",
            last: "Selden"
        };
    };
    IfScenario.prototype.test = function (render) {
        var result = render().outerHTML;
        if (result !== '<div>Kris Selden</div>') {
            throw new Error("Invalid render: " + result);
        }
    };
    return IfScenario;
}(bench_1.TemplateBenchmarkScenario));
var InlineIfScenario = (function (_super) {
    __extends(InlineIfScenario, _super);
    function InlineIfScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'inline if';
        _this.description = 'two `inline-if` helpers';
        return _this;
    }
    InlineIfScenario.prototype.template = function () {
        return '{{inline-if truthy first "empty"}} {{inline-if falsy "empty" last}}';
    };
    return InlineIfScenario;
}(IfScenario));
var BlockIfScenario = (function (_super) {
    __extends(BlockIfScenario, _super);
    function BlockIfScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = 'block if';
        _this.description = 'two `{{#if}}` blocks';
        return _this;
    }
    BlockIfScenario.prototype.template = function () {
        return '{{#if truthy}}{{first}}{{else}}empty{{/if}} {{#if falsy}}empty{{else}}{{last}}{{/if}}';
    };
    return BlockIfScenario;
}(IfScenario));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = [
    InlineIfScenario,
    BlockIfScenario
];
