// glimmer-benchmarks/lib/benchmarks/partials.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var bench_1 = require("../bench");
var StaticPartialScenario = (function (_super) {
    __extends(StaticPartialScenario, _super);
    function StaticPartialScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = "static partial";
        _this.description = "`{{partial 'greeting'}}`";
        return _this;
    }
    StaticPartialScenario.prototype.start = function () {
        this.glimmerEnv.registerPartial('greeting', "Hi {{person1.name}} from a partial");
        _super.prototype.start.call(this);
    };
    StaticPartialScenario.prototype.template = function () {
        return "[{{partial 'greeting'}}]";
    };
    StaticPartialScenario.prototype.renderContext = function () {
        return {
            person1: { name: 'Alex Joyce' }
        };
    };
    StaticPartialScenario.prototype.test = function (render) {
        var result = render().outerHTML;
        if (result !== '<div>[Hi Alex Joyce from a partial]</div>') {
            throw new Error("Invalid render: " + result);
        }
    };
    return StaticPartialScenario;
}(bench_1.TemplateBenchmarkScenario));
var DynamicPartialScenario = (function (_super) {
    __extends(DynamicPartialScenario, _super);
    function DynamicPartialScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = "dynamic partial";
        _this.description = "`{{partial partialName}}`";
        return _this;
    }
    DynamicPartialScenario.prototype.start = function () {
        this.glimmerEnv.registerPartial('greeting', "Hi {{person1.name}} from a partial");
        _super.prototype.start.call(this);
    };
    DynamicPartialScenario.prototype.template = function () {
        return "[{{partial partialName}}]";
    };
    DynamicPartialScenario.prototype.renderContext = function () {
        return {
            partialName: 'greeting',
            person1: { name: 'Ben Joyce' }
        };
    };
    DynamicPartialScenario.prototype.test = function (render) {
        var result = render().outerHTML;
        if (result !== '<div>[Hi Ben Joyce from a partial]</div>') {
            throw new Error("Invalid render: " + result);
        }
    };
    return DynamicPartialScenario;
}(bench_1.TemplateBenchmarkScenario));
var BaselineScenario = (function (_super) {
    __extends(BaselineScenario, _super);
    function BaselineScenario() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = "baseline";
        _this.description = "a template with no partial";
        return _this;
    }
    BaselineScenario.prototype.template = function () {
        return "[Hi {{person1.name}} from a template]";
    };
    BaselineScenario.prototype.renderContext = function () {
        return {
            person1: { name: 'Sarah Galvin' }
        };
    };
    BaselineScenario.prototype.test = function (render) {
        var result = render().outerHTML;
        if (result !== '<div>[Hi Sarah Galvin from a template]</div>') {
            throw new Error("Invalid render: " + result);
        }
    };
    return BaselineScenario;
}(bench_1.TemplateBenchmarkScenario));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = [
    StaticPartialScenario,
    DynamicPartialScenario,
    BaselineScenario
];
