// glimmer-engine/packages/@glimmer/object-reference/dist/commonjs/lib/references/path.js
"use strict";
const utils_1 = require("../utils");
const util_1 = require("@glimmer/util");
const meta_1 = require("../meta");
const descriptors_1 = require("./descriptors");
const reference_1 = require("@glimmer/reference");
class UnchainFromPath {
    constructor(set, child) {
        this.set = set;
        this.child = child;
    }
    destroy() {
        this.set.delete(this.child);
    }
}
class PathReference {
    constructor(parent, property) {
        this.cache = utils_1.EMPTY_CACHE;
        this.inner = null;
        this.chains = null;
        this.lastParentValue = utils_1.EMPTY_CACHE;
        this._guid = null;
        this.tag = reference_1.VOLATILE_TAG;
        this.parent = parent;
        this.property = property;
    }
    value() {
        let { lastParentValue, property, inner } = this;
        let parentValue = this._parentValue();
        if (parentValue === null || parentValue === undefined) {
            return (this.cache = undefined);
        }
        if (lastParentValue === parentValue) {
            inner = this.inner;
        }
        else {
            let ReferenceType = typeof parentValue === 'object' ? meta_1.default.for(parentValue).referenceTypeFor(property) : descriptors_1.PropertyReference;
            inner = this.inner = new ReferenceType(parentValue, property, this);
        }
        // if (typeof parentValue === 'object') {
        //   Meta.for(parentValue).addReference(property, this);
        // }
        return (this.cache = inner.value());
    }
    get(prop) {
        let chains = this._getChains();
        if (prop in chains)
            return chains[prop];
        return (chains[prop] = new PathReference(this, prop));
    }
    label() {
        return '[reference Direct]';
    }
    _getChains() {
        if (this.chains)
            return this.chains;
        return (this.chains = util_1.dict());
    }
    _parentValue() {
        let parent = this.parent.value();
        this.lastParentValue = parent;
        return parent;
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PathReference;
exports.PathReference = PathReference;
