// glimmer-engine/packages/glimmer-runtime/dist/commonjs/index.js
"use strict";
require("./lib/bootstrap");
var template_1 = require("./lib/template");
exports.templateFactory = template_1.default;
var references_1 = require("./lib/references");
exports.NULL_REFERENCE = references_1.NULL_REFERENCE;
exports.UNDEFINED_REFERENCE = references_1.UNDEFINED_REFERENCE;
exports.PrimitiveReference = references_1.PrimitiveReference;
exports.ConditionalReference = references_1.ConditionalReference;
var builder_1 = require("./lib/compiled/opcodes/builder");
exports.OpcodeBuilderDSL = builder_1.default;
var compiler_1 = require("./lib/compiler");
exports.compileLayout = compiler_1.compileLayout;
var blocks_1 = require("./lib/compiled/blocks");
exports.CompiledBlock = blocks_1.CompiledBlock;
exports.CompiledProgram = blocks_1.CompiledProgram;
var attribute_managers_1 = require("./lib/dom/attribute-managers");
exports.IAttributeManager = attribute_managers_1.AttributeManager;
exports.AttributeManager = attribute_managers_1.AttributeManager;
exports.PropertyManager = attribute_managers_1.PropertyManager;
exports.INPUT_VALUE_PROPERTY_MANAGER = attribute_managers_1.INPUT_VALUE_PROPERTY_MANAGER;
exports.defaultManagers = attribute_managers_1.defaultManagers;
exports.defaultAttributeManagers = attribute_managers_1.defaultAttributeManagers;
exports.defaultPropertyManagers = attribute_managers_1.defaultPropertyManagers;
exports.readDOMAttr = attribute_managers_1.readDOMAttr;
var content_1 = require("./lib/compiled/opcodes/content");
exports.normalizeTextValue = content_1.normalizeTextValue;
var expressions_1 = require("./lib/compiled/expressions");
exports.CompiledExpression = expressions_1.CompiledExpression;
var args_1 = require("./lib/compiled/expressions/args");
exports.CompiledArgs = args_1.CompiledArgs;
exports.CompiledNamedArgs = args_1.CompiledNamedArgs;
exports.CompiledPositionalArgs = args_1.CompiledPositionalArgs;
exports.EvaluatedArgs = args_1.EvaluatedArgs;
exports.EvaluatedNamedArgs = args_1.EvaluatedNamedArgs;
exports.EvaluatedPositionalArgs = args_1.EvaluatedPositionalArgs;
var get_dynamic_var_1 = require("./lib/helpers/get-dynamic-var");
exports.getDynamicVar = get_dynamic_var_1.default;
var functions_1 = require("./lib/syntax/functions");
exports.BlockMacros = functions_1.Blocks;
exports.InlineMacros = functions_1.Inlines;
exports.compileArgs = functions_1.compileArgs;
var scanner_1 = require("./lib/scanner");
exports.BaselineSyntax = scanner_1.BaselineSyntax;
exports.Layout = scanner_1.Layout;
var vm_1 = require("./lib/vm");
exports.UpdatingVM = vm_1.UpdatingVM;
exports.RenderResult = vm_1.RenderResult;
var upsert_1 = require("./lib/upsert");
exports.isSafeString = upsert_1.isSafeString;
var environment_1 = require("./lib/environment");
exports.Scope = environment_1.Scope;
exports.Environment = environment_1.default;
var partial_1 = require("./lib/partial");
exports.PartialDefinition = partial_1.PartialDefinition;
var interfaces_1 = require("./lib/component/interfaces");
exports.ComponentDefinition = interfaces_1.ComponentDefinition;
exports.isComponentDefinition = interfaces_1.isComponentDefinition;
var helper_1 = require("./lib/dom/helper");
exports.DOMChanges = helper_1.default;
exports.IDOMChanges = helper_1.DOMChanges;
exports.DOMTreeConstruction = helper_1.DOMTreeConstruction;
exports.isWhitespace = helper_1.isWhitespace;
exports.insertHTMLBefore = helper_1.insertHTMLBefore;
const Simple = require("./lib/dom/interfaces");
exports.Simple = Simple;
var builder_2 = require("./lib/builder");
exports.ElementStack = builder_2.ElementStack;
var bounds_1 = require("./lib/bounds");
exports.ConcreteBounds = bounds_1.ConcreteBounds;
