// @glimmer/runtime/lib/compiled/expressions/args.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var positional_args_1 = require("./positional-args");
exports.COMPILED_EMPTY_POSITIONAL_ARGS = positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS;
exports.CompiledPositionalArgs = positional_args_1.CompiledPositionalArgs;
exports.EvaluatedPositionalArgs = positional_args_1.EvaluatedPositionalArgs;
var named_args_1 = require("./named-args");
exports.COMPILED_EMPTY_NAMED_ARGS = named_args_1.COMPILED_EMPTY_NAMED_ARGS;
exports.CompiledNamedArgs = named_args_1.CompiledNamedArgs;
exports.EvaluatedNamedArgs = named_args_1.EvaluatedNamedArgs;
var reference_1 = require("@glimmer/reference");
exports.EMPTY_BLOCKS = {
    default: null,
    inverse: null
};
var CompiledArgs = (function () {
    function CompiledArgs(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.type = "compiled-args";
    }
    CompiledArgs.create = function (positional, named, blocks) {
        if (positional === positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS && named === named_args_1.COMPILED_EMPTY_NAMED_ARGS && blocks === exports.EMPTY_BLOCKS) {
            return this.empty();
        }
        else {
            return new this(positional, named, blocks);
        }
    };
    CompiledArgs.empty = function () {
        return exports.COMPILED_EMPTY_ARGS;
    };
    CompiledArgs.prototype.evaluate = function (vm) {
        var _a = this, positional = _a.positional, named = _a.named, blocks = _a.blocks;
        return EvaluatedArgs.create(positional.evaluate(vm), named.evaluate(vm), blocks);
    };
    return CompiledArgs;
}());
exports.CompiledArgs = CompiledArgs;
exports.COMPILED_EMPTY_ARGS = new ((function (_super) {
    __extends(class_1, _super);
    function class_1() {
        return _super.call(this, positional_args_1.COMPILED_EMPTY_POSITIONAL_ARGS, named_args_1.COMPILED_EMPTY_NAMED_ARGS, exports.EMPTY_BLOCKS) || this;
    }
    class_1.prototype.evaluate = function (_vm) {
        return EMPTY_EVALUATED_ARGS;
    };
    return class_1;
}(CompiledArgs)));
var EvaluatedArgs = (function () {
    function EvaluatedArgs(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.tag = reference_1.combineTagged([positional, named]);
    }
    EvaluatedArgs.empty = function () {
        return EMPTY_EVALUATED_ARGS;
    };
    EvaluatedArgs.create = function (positional, named, blocks) {
        return new this(positional, named, blocks);
    };
    EvaluatedArgs.positional = function (values, blocks) {
        if (blocks === void 0) { blocks = exports.EMPTY_BLOCKS; }
        return new this(positional_args_1.EvaluatedPositionalArgs.create(values), named_args_1.EVALUATED_EMPTY_NAMED_ARGS, blocks);
    };
    EvaluatedArgs.named = function (map, blocks) {
        if (blocks === void 0) { blocks = exports.EMPTY_BLOCKS; }
        return new this(positional_args_1.EVALUATED_EMPTY_POSITIONAL_ARGS, named_args_1.EvaluatedNamedArgs.create(map), blocks);
    };
    return EvaluatedArgs;
}());
exports.EvaluatedArgs = EvaluatedArgs;
var EMPTY_EVALUATED_ARGS = new EvaluatedArgs(positional_args_1.EVALUATED_EMPTY_POSITIONAL_ARGS, named_args_1.EVALUATED_EMPTY_NAMED_ARGS, exports.EMPTY_BLOCKS);
