// @glimmer/runtime/lib/compiled/expressions/concat.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var reference_1 = require("@glimmer/reference");
var CompiledConcat = (function () {
    function CompiledConcat(parts) {
        this.parts = parts;
        this.type = "concat";
    }
    CompiledConcat.prototype.evaluate = function (vm) {
        var parts = new Array(this.parts.length);
        for (var i = 0; i < this.parts.length; i++) {
            parts[i] = this.parts[i].evaluate(vm);
        }
        return new ConcatReference(parts);
    };
    CompiledConcat.prototype.toJSON = function () {
        return "concat(" + this.parts.map(function (expr) { return expr.toJSON(); }).join(", ") + ")";
    };
    return CompiledConcat;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledConcat;
var ConcatReference = (function (_super) {
    __extends(ConcatReference, _super);
    function ConcatReference(parts) {
        var _this = _super.call(this) || this;
        _this.parts = parts;
        _this.tag = reference_1.combineTagged(parts);
        return _this;
    }
    ConcatReference.prototype.compute = function () {
        var parts = new Array();
        for (var i = 0; i < this.parts.length; i++) {
            var value = this.parts[i].value();
            if (value !== null && value !== undefined) {
                parts[i] = castToString(value);
            }
        }
        if (parts.length > 0) {
            return parts.join('');
        }
        return null;
    };
    return ConcatReference;
}(reference_1.CachedReference));
exports.ConcatReference = ConcatReference;
function castToString(value) {
    if (typeof value['toString'] !== 'function') {
        return '';
    }
    return String(value);
}
