// @glimmer/runtime/lib/compiled/expressions/has-block.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var expressions_1 = require("../expressions");
var references_1 = require("../../references");
var CompiledHasBlock = (function (_super) {
    __extends(CompiledHasBlock, _super);
    function CompiledHasBlock(inner) {
        var _this = _super.call(this) || this;
        _this.inner = inner;
        _this.type = "has-block";
        return _this;
    }
    CompiledHasBlock.prototype.evaluate = function (vm) {
        var block = this.inner.evaluate(vm);
        return references_1.PrimitiveReference.create(!!block);
    };
    CompiledHasBlock.prototype.toJSON = function () {
        return "has-block(" + this.inner.toJSON() + ")";
    };
    return CompiledHasBlock;
}(expressions_1.CompiledExpression));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledHasBlock;
var CompiledHasBlockParams = (function (_super) {
    __extends(CompiledHasBlockParams, _super);
    function CompiledHasBlockParams(inner) {
        var _this = _super.call(this) || this;
        _this.inner = inner;
        _this.type = "has-block-params";
        return _this;
    }
    CompiledHasBlockParams.prototype.evaluate = function (vm) {
        var block = this.inner.evaluate(vm);
        var hasLocals = block && block.symbolTable.getSymbols().locals;
        return references_1.PrimitiveReference.create(!!hasLocals);
    };
    CompiledHasBlockParams.prototype.toJSON = function () {
        return "has-block-params(" + this.inner.toJSON() + ")";
    };
    return CompiledHasBlockParams;
}(expressions_1.CompiledExpression));
exports.CompiledHasBlockParams = CompiledHasBlockParams;
var CompiledGetBlockBySymbol = (function () {
    function CompiledGetBlockBySymbol(symbol, debug) {
        this.symbol = symbol;
        this.debug = debug;
    }
    CompiledGetBlockBySymbol.prototype.evaluate = function (vm) {
        return vm.scope().getBlock(this.symbol);
    };
    CompiledGetBlockBySymbol.prototype.toJSON = function () {
        return "get-block($" + this.symbol + "(" + this.debug + "))";
    };
    return CompiledGetBlockBySymbol;
}());
exports.CompiledGetBlockBySymbol = CompiledGetBlockBySymbol;
var CompiledInPartialGetBlock = (function () {
    function CompiledInPartialGetBlock(symbol, name) {
        this.symbol = symbol;
        this.name = name;
    }
    CompiledInPartialGetBlock.prototype.evaluate = function (vm) {
        var _a = this, symbol = _a.symbol, name = _a.name;
        var args = vm.scope().getPartialArgs(symbol);
        return args.blocks[name];
    };
    CompiledInPartialGetBlock.prototype.toJSON = function () {
        return "get-block($" + this.symbol + "($ARGS)." + this.name + "))";
    };
    return CompiledInPartialGetBlock;
}());
exports.CompiledInPartialGetBlock = CompiledInPartialGetBlock;
