// @glimmer/runtime/lib/compiled/expressions/lookups.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var expressions_1 = require("../expressions");
var reference_1 = require("@glimmer/reference");
var CompiledLookup = (function (_super) {
    __extends(CompiledLookup, _super);
    function CompiledLookup(base, path) {
        var _this = _super.call(this) || this;
        _this.base = base;
        _this.path = path;
        _this.type = "lookup";
        return _this;
    }
    CompiledLookup.create = function (base, path) {
        if (path.length === 0) {
            return base;
        }
        else {
            return new this(base, path);
        }
    };
    CompiledLookup.prototype.evaluate = function (vm) {
        var _a = this, base = _a.base, path = _a.path;
        return reference_1.referenceFromParts(base.evaluate(vm), path);
    };
    CompiledLookup.prototype.toJSON = function () {
        return this.base.toJSON() + "." + this.path.join('.');
    };
    return CompiledLookup;
}(expressions_1.CompiledExpression));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledLookup;
var CompiledSelf = (function (_super) {
    __extends(CompiledSelf, _super);
    function CompiledSelf() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CompiledSelf.prototype.evaluate = function (vm) {
        return vm.getSelf();
    };
    CompiledSelf.prototype.toJSON = function () {
        return 'self';
    };
    return CompiledSelf;
}(expressions_1.CompiledExpression));
exports.CompiledSelf = CompiledSelf;
var CompiledSymbol = (function (_super) {
    __extends(CompiledSymbol, _super);
    function CompiledSymbol(symbol, debug) {
        var _this = _super.call(this) || this;
        _this.symbol = symbol;
        _this.debug = debug;
        return _this;
    }
    CompiledSymbol.prototype.evaluate = function (vm) {
        return vm.referenceForSymbol(this.symbol);
    };
    CompiledSymbol.prototype.toJSON = function () {
        return "$" + this.symbol + "(" + this.debug + ")";
    };
    return CompiledSymbol;
}(expressions_1.CompiledExpression));
exports.CompiledSymbol = CompiledSymbol;
var CompiledInPartialName = (function (_super) {
    __extends(CompiledInPartialName, _super);
    function CompiledInPartialName(symbol, name) {
        var _this = _super.call(this) || this;
        _this.symbol = symbol;
        _this.name = name;
        return _this;
    }
    CompiledInPartialName.prototype.evaluate = function (vm) {
        var _a = this, symbol = _a.symbol, name = _a.name;
        var args = vm.scope().getPartialArgs(symbol);
        return args.named.get(name);
    };
    CompiledInPartialName.prototype.toJSON = function () {
        return "$" + this.symbol + "($ARGS)." + this.name;
    };
    return CompiledInPartialName;
}(expressions_1.CompiledExpression));
exports.CompiledInPartialName = CompiledInPartialName;
