// @glimmer/runtime/lib/compiled/expressions/positional-args.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var references_1 = require("../../references");
var utils_1 = require("../../utils");
var reference_1 = require("@glimmer/reference");
var CompiledPositionalArgs = (function () {
    function CompiledPositionalArgs(values) {
        this.values = values;
        this.length = values.length;
    }
    CompiledPositionalArgs.create = function (values) {
        if (values.length) {
            return new this(values);
        }
        else {
            return exports.COMPILED_EMPTY_POSITIONAL_ARGS;
        }
    };
    CompiledPositionalArgs.empty = function () {
        return exports.COMPILED_EMPTY_POSITIONAL_ARGS;
    };
    CompiledPositionalArgs.prototype.evaluate = function (vm) {
        var _a = this, values = _a.values, length = _a.length;
        var references = new Array(length);
        for (var i = 0; i < length; i++) {
            references[i] = values[i].evaluate(vm);
        }
        return EvaluatedPositionalArgs.create(references);
    };
    CompiledPositionalArgs.prototype.toJSON = function () {
        return "[" + this.values.map(function (value) { return value.toJSON(); }).join(", ") + "]";
    };
    return CompiledPositionalArgs;
}());
exports.CompiledPositionalArgs = CompiledPositionalArgs;
exports.COMPILED_EMPTY_POSITIONAL_ARGS = new ((function (_super) {
    __extends(class_1, _super);
    function class_1() {
        return _super.call(this, utils_1.EMPTY_ARRAY) || this;
    }
    class_1.prototype.evaluate = function (_vm) {
        return exports.EVALUATED_EMPTY_POSITIONAL_ARGS;
    };
    class_1.prototype.toJSON = function () {
        return "<EMPTY>";
    };
    return class_1;
}(CompiledPositionalArgs)));
var EvaluatedPositionalArgs = (function () {
    function EvaluatedPositionalArgs(values) {
        this.values = values;
        this.tag = reference_1.combineTagged(values);
        this.length = values.length;
    }
    EvaluatedPositionalArgs.create = function (values) {
        return new this(values);
    };
    EvaluatedPositionalArgs.empty = function () {
        return exports.EVALUATED_EMPTY_POSITIONAL_ARGS;
    };
    EvaluatedPositionalArgs.prototype.at = function (index) {
        var _a = this, values = _a.values, length = _a.length;
        return (index < length) ? values[index] : references_1.UNDEFINED_REFERENCE;
    };
    EvaluatedPositionalArgs.prototype.value = function () {
        var _a = this, values = _a.values, length = _a.length;
        var ret = new Array(length);
        for (var i = 0; i < length; i++) {
            ret[i] = values[i].value();
        }
        return ret;
    };
    return EvaluatedPositionalArgs;
}());
exports.EvaluatedPositionalArgs = EvaluatedPositionalArgs;
exports.EVALUATED_EMPTY_POSITIONAL_ARGS = new ((function (_super) {
    __extends(class_2, _super);
    function class_2() {
        return _super.call(this, utils_1.EMPTY_ARRAY) || this;
    }
    class_2.prototype.at = function () {
        return references_1.UNDEFINED_REFERENCE;
    };
    class_2.prototype.value = function () {
        return this.values;
    };
    return class_2;
}(EvaluatedPositionalArgs)));
