// glimmer-engine/dist/named-amd/@glimmer/object-reference/index.js
enifed("@glimmer/object-reference/index", ["exports", "@glimmer/object-reference/lib/references/descriptors", "@glimmer/object-reference/lib/meta", "@glimmer/object-reference/lib/object", "@glimmer/object-reference/lib/types", "@glimmer/object-reference/lib/references/path", "@glimmer/object-reference/lib/references/root", "@glimmer/reference"], function (exports, _glimmerObjectReferenceLibReferencesDescriptors, _glimmerObjectReferenceLibMeta, _glimmerObjectReferenceLibObject, _glimmerObjectReferenceLibTypes, _glimmerObjectReferenceLibReferencesPath, _glimmerObjectReferenceLibReferencesRoot, _glimmerReference) {
  "use strict";

  function _interopExportWildcard(obj, defaults) { var newObj = defaults({}, obj); delete newObj["default"]; return newObj; }

  function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

  exports.ComputedReferenceBlueprint = _glimmerObjectReferenceLibReferencesDescriptors.ComputedReferenceBlueprint;
  exports.InnerReferenceFactory = _glimmerObjectReferenceLibReferencesDescriptors.InnerReferenceFactory;
  exports.PropertyReference = _glimmerObjectReferenceLibReferencesDescriptors.PropertyReference;
  exports.CLASS_META = _glimmerObjectReferenceLibMeta.CLASS_META;
  exports.Meta = _glimmerObjectReferenceLibMeta.default;
  exports.metaFor = _glimmerObjectReferenceLibMeta.metaFor;
  exports.setProperty = _glimmerObjectReferenceLibObject.setProperty;
  exports.notifyProperty = _glimmerObjectReferenceLibObject.notifyProperty;

  _defaults(exports, _interopExportWildcard(_glimmerObjectReferenceLibTypes, _defaults));

  exports.ObjectReference = _glimmerObjectReferenceLibReferencesPath.default;
  exports.UpdatableReference = _glimmerObjectReferenceLibReferencesRoot.default;
  exports.ConstReference = _glimmerReference.ConstReference;
  exports.isConst = _glimmerReference.isConst;
});
