// glimmer-engine/dist/es6/@glimmer/object-model/lib/object.js
import { GlimmerDescriptor, ValueBlueprint } from './blueprint';
export class GlimmerObjectBase {
}
export default class GlimmerObject {
    static create(properties) {
        return new this(properties);
    }
    static extend(extensions) {
        // This method intentionally uses internal typecasts to convince TypeScript
        // that what we're doing is legitimate. The real action of this method is in
        // its signature (and the signature of GlimmerClass).
        let sub = class extends this {
        };
        if (extensions !== undefined) {
            let blueprints = Object.keys(extensions).map(k => {
                let extension = extensions[k];
                if (extension instanceof GlimmerDescriptor) {
                    return extension.blueprint(k);
                }
                else {
                    return new ValueBlueprint(k, extensions[k]);
                }
            });
            blueprints.forEach(b => b.define(sub.prototype));
        }
        return sub;
    }
    constructor(properties) {
        if (properties !== undefined) {
            Object.assign(this, properties);
        }
    }
}
/**
 * Adapter for TypeScript.
 *
 * A shim library could just declare the variable as the type
 * GlimmerClass<FirstInstance>. The key point is that shimming
 * between the two worlds requires writing a static interface
 * definition, as TypeScript isn't willing to treat the dynamic
 * intersection as a valid constructor return type.
 *
 * This function is a noop and is used like this:
 *
 * ```ts
 * // in one file
 * let Person = GlimmerObject.extend({
 *   named: 'Dan'
 * });
 *
 *
 * // in a TypeScript wrapper file
 * import { Person as OriginalPerson } from 'original-definition';
 *
 * interface PersonInstance {
 *   named: string
 * }
 *
 * export let Person = classof<PersonInstance>(OriginalPerson);
 *
 * // in a subclassing file
 * import { Person } from 'wrapped-definition';
 *
 * class FancyPerson extends Person {
 *   salutation: string;
 *
 *   fullName() {
 *     return `${this.salutation} ${this.named}`;
 *   }
 * }
 *
 * FancyPerson.create({ name: 'Tom Dale', salutation: 'Mr.' }).fullName();
 * // typechecks and returns 'Mr. Tom Dale'
 * ```
*/
export function classof(klass) {
    return klass;
}
