// glimmer-engine/dist/named-amd/@glimmer/runtime/lib/syntax/specialize.js
enifed('@glimmer/runtime/lib/syntax/specialize', ['exports', '@glimmer/wire-format', '@glimmer/util'], function (exports, _glimmerWireFormat, _glimmerUtil) {
    'use strict';

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var Specialize = (function () {
        function Specialize() {
            _classCallCheck(this, Specialize);

            this.names = _glimmerUtil.dict();
            this.funcs = [];
        }

        Specialize.prototype.add = function add(name, func) {
            this.funcs.push(func);
            this.names[name] = this.funcs.length - 1;
        };

        Specialize.prototype.specialize = function specialize(sexp, table) {
            var name = sexp[0];
            var index = this.names[name];
            if (index === undefined) return sexp;
            var func = this.funcs[index];
            _glimmerUtil.assert(!!func, 'expected a specialization for ' + sexp[0]);
            return func(sexp, table);
        };

        return Specialize;
    })();

    exports.Specialize = Specialize;
    var SPECIALIZE = new Specialize();
    exports.SPECIALIZE = SPECIALIZE;
    var E = _glimmerWireFormat.Expressions;
    SPECIALIZE.add('append', function (sexp, _symbolTable) {
        var path = sexp[1];
        if (Array.isArray(path) && (E.isUnknown(path) || E.isGet(path))) {
            if (path[1].length !== 1) {
                return ['unoptimized-append', sexp[1], sexp[2]];
            }
        }
        return ['optimized-append', sexp[1], sexp[2]];
    });
    SPECIALIZE.add('dynamic-attr', function (sexp, _symbolTable) {
        return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], false];
    });
    SPECIALIZE.add('trusting-attr', function (sexp, _symbolTable) {
        return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], true];
    });
    SPECIALIZE.add('partial', function (sexp, _table) {
        var expression = sexp[1];
        if (typeof expression === 'string') {
            return ['static-partial', expression];
        } else {
            return ['dynamic-partial', expression];
        }
    });
});
