// glimmer-engine/packages/glimmer-object-reference/dist/commonjs/lib/references/descriptors.js
"use strict";
const meta_1 = require("../meta");
const glimmer_reference_1 = require("glimmer-reference");
class PropertyReference {
    constructor(object, property, outer) {
        this.tag = glimmer_reference_1.VOLATILE_TAG;
        this.object = object;
        this.property = property;
    }
    value() { return this.object[this.property]; }
    label() {
        return '[reference Property]';
    }
}
exports.PropertyReference = PropertyReference;
function ComputedReferenceBlueprint(property, dependencies) {
    return class ComputedReference {
        constructor(object, property, outer) {
            this.installed = false;
            this.tag = glimmer_reference_1.VOLATILE_TAG;
            this.object = object;
            this.property = property;
            this.dependencies = dependencies;
            this.outer = outer;
        }
        value() {
            if (!this.installed) {
                let root = meta_1.default.for(this.object).root();
                this.dependencies.forEach(dep => {
                    let ref = root.referenceFromParts(dep);
                    ref.value();
                });
                this.installed = true;
            }
            return this.object[this.property];
        }
        label() {
            return '[reference Computed]';
        }
    };
}
exports.ComputedReferenceBlueprint = ComputedReferenceBlueprint;
