// glimmer-engine/packages/glimmer-object-reference/dist/commonjs/lib/references/root.js
"use strict";
const glimmer_util_1 = require("glimmer-util");
const path_1 = require("./path");
const glimmer_reference_1 = require("glimmer-reference");
class RootReference {
    constructor(object) {
        this.chains = glimmer_util_1.dict();
        this.tag = glimmer_reference_1.VOLATILE_TAG;
        this.object = object;
    }
    value() { return this.object; }
    update(object) {
        this.object = object;
        // this.notify();
    }
    get(prop) {
        let chains = this.chains;
        if (prop in chains)
            return chains[prop];
        return (chains[prop] = new path_1.PathReference(this, prop));
    }
    chainFor(prop) {
        let chains = this.chains;
        if (prop in chains)
            return chains[prop];
        return null;
    }
    path(string) {
        return string.split('.').reduce((ref, part) => ref.get(part), this);
    }
    referenceFromParts(parts) {
        return parts.reduce((ref, part) => ref.get(part), this);
    }
    label() {
        return '[reference Root]';
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RootReference;
