// glimmer-engine/packages/@glimmer/object-model/dist/commonjs/lib/reference.js
"use strict";
const meta_1 = require("./meta");
const reference_1 = require("@glimmer/reference");
const util_1 = require("@glimmer/util");
if (util_1.HAS_NATIVE_WEAKMAP) {
    const META = new WeakMap();
    const CLASS_META = new WeakMap();
    exports.classMeta = function _classMetaNative(object) {
        let m = CLASS_META.get(object);
        if (m === undefined) {
            m = new meta_1.ClassMeta();
            CLASS_META.set(object, m);
        }
        return m;
    };
    exports.meta = function _metaNative(object) {
        let m = META.get(object);
        if (m === undefined) {
            m = new meta_1.default();
            META.set(object, m);
        }
        return m;
    };
}
else {
    const GLIMMER_META = 'META__glimmer__1484170086860394543206811';
    const GLIMMER_CLASS_META = 'CLASS_META__glimmer__14841708559821468834708062';
    exports.classMeta = function _classMetaFaux(object) {
        let m = object[GLIMMER_CLASS_META];
        if (m === undefined) {
            m = new meta_1.ClassMeta();
            object[GLIMMER_CLASS_META] = m;
        }
        return m;
    };
    exports.meta = function _metaFaux(object) {
        let m = object[GLIMMER_META];
        if (m === undefined) {
            m = new meta_1.default();
            object[GLIMMER_META] = m;
        }
        return m;
    };
}
function set(object, key, value) {
    object[key] = value;
    exports.meta(object).dirty(key);
}
exports.set = set;
function root(object) {
    return new VersionedRootReference(object);
}
exports.root = root;
class VersionedRootReference {
    constructor(inner) {
        this.inner = inner;
        this.tag = reference_1.CONSTANT_TAG;
    }
    value() {
        return this.inner;
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
exports.VersionedRootReference = VersionedRootReference;
class VersionedObjectReference {
    constructor(parent, key) {
        this.parent = parent;
        this.key = key;
        this.tag = reference_1.CURRENT_TAG;
    }
    value() {
        let { parent, key } = this;
        let parentObject = this.parent.value();
        let computed = exports.classMeta(Object.getPrototypeOf(parentObject)).getComputed(key);
        let tags = [exports.meta(parentObject).tag(key)];
        if (computed) {
            tags.push(...computed.dependentKeys.map(key => path(this, key).tag));
        }
        this.tag = reference_1.combine(tags);
        return parentObject[key];
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
exports.VersionedObjectReference = VersionedObjectReference;
function path(parent, key) {
    return key.split('.').reduce((ref, part) => ref.get(part), parent);
}
