// @glimmer/util/lib/collections.js
"use strict";
var guid_1 = require("./guid");
var proto = Object.create(null, {
    // without this, we will always still end up with (new
    // EmptyObject()).constructor === Object
    constructor: {
        value: undefined,
        enumerable: false,
        writable: true
    }
});
function EmptyObject() { }
EmptyObject.prototype = proto;
function dict() {
    // let d = Object.create(null);
    // d.x = 1;
    // delete d.x;
    // return d;
    return new EmptyObject();
}
exports.dict = dict;
var DictSet = (function () {
    function DictSet() {
        this.dict = dict();
    }
    DictSet.prototype.add = function (obj) {
        if (typeof obj === 'string')
            this.dict[obj] = obj;
        else
            this.dict[guid_1.ensureGuid(obj)] = obj;
        return this;
    };
    DictSet.prototype.delete = function (obj) {
        if (typeof obj === 'string')
            delete this.dict[obj];
        else if (obj._guid)
            delete this.dict[obj._guid];
    };
    DictSet.prototype.forEach = function (callback) {
        var dict = this.dict;
        Object.keys(dict).forEach(function (key) { return callback(dict[key]); });
    };
    DictSet.prototype.toArray = function () {
        return Object.keys(this.dict);
    };
    return DictSet;
}());
exports.DictSet = DictSet;
var Stack = (function () {
    function Stack() {
        this.stack = [];
        this.current = null;
    }
    Stack.prototype.toArray = function () {
        return this.stack;
    };
    Stack.prototype.push = function (item) {
        this.current = item;
        this.stack.push(item);
    };
    Stack.prototype.pop = function () {
        var item = this.stack.pop();
        var len = this.stack.length;
        this.current = len === 0 ? null : this.stack[len - 1];
        return item === undefined ? null : item;
    };
    Stack.prototype.isEmpty = function () {
        return this.stack.length === 0;
    };
    return Stack;
}());
exports.Stack = Stack;
