// @glimmer/util/lib/logger.js
"use strict";
var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["Trace"] = 0] = "Trace";
    LogLevel[LogLevel["Debug"] = 1] = "Debug";
    LogLevel[LogLevel["Warn"] = 2] = "Warn";
    LogLevel[LogLevel["Error"] = 3] = "Error";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
var NullConsole = (function () {
    function NullConsole() {
    }
    NullConsole.prototype.log = function (_message) { };
    NullConsole.prototype.warn = function (_message) { };
    NullConsole.prototype.error = function (_message) { };
    NullConsole.prototype.trace = function () { };
    return NullConsole;
}());
var ALWAYS;
var Logger = (function () {
    function Logger(_a) {
        var console = _a.console, level = _a.level;
        this.f = ALWAYS;
        this.force = ALWAYS;
        this.console = console;
        this.level = level;
    }
    Logger.prototype.skipped = function (level) {
        return level < this.level;
    };
    Logger.prototype.trace = function (message, _a) {
        var _b = (_a === void 0 ? {} : _a).stackTrace, stackTrace = _b === void 0 ? false : _b;
        if (this.skipped(LogLevel.Trace))
            return;
        this.console.log(message);
        if (stackTrace)
            this.console.trace();
    };
    Logger.prototype.debug = function (message, _a) {
        var _b = (_a === void 0 ? {} : _a).stackTrace, stackTrace = _b === void 0 ? false : _b;
        if (this.skipped(LogLevel.Debug))
            return;
        this.console.log(message);
        if (stackTrace)
            this.console.trace();
    };
    Logger.prototype.warn = function (message, _a) {
        var _b = (_a === void 0 ? {} : _a).stackTrace, stackTrace = _b === void 0 ? false : _b;
        if (this.skipped(LogLevel.Warn))
            return;
        this.console.warn(message);
        if (stackTrace)
            this.console.trace();
    };
    Logger.prototype.error = function (message) {
        if (this.skipped(LogLevel.Error))
            return;
        this.console.error(message);
    };
    return Logger;
}());
exports.Logger = Logger;
var _console = (typeof console === 'undefined') ? new NullConsole() : console;
ALWAYS = new Logger({ console: _console, level: LogLevel.Trace });
var LOG_LEVEL = LogLevel.Warn;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = new Logger({ console: _console, level: LOG_LEVEL });
