// @glimmer/object/lib/descriptors.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var mixin_1 = require("./mixin");
var computed_1 = require("./computed");
var AliasMethodDescriptor = (function (_super) {
    __extends(AliasMethodDescriptor, _super);
    function AliasMethodDescriptor(name) {
        var _this = _super.call(this) || this;
        _this.name = name;
        return _this;
    }
    AliasMethodDescriptor.prototype.define = function (target, key, home) {
        var name = this.name;
        Object.defineProperty(target, key, {
            enumerable: true,
            configurable: true,
            get: function () {
                return this[name];
            }
        });
    };
    return AliasMethodDescriptor;
}(mixin_1.Descriptor));
var AliasMethodBlueprint = (function (_super) {
    __extends(AliasMethodBlueprint, _super);
    function AliasMethodBlueprint(name) {
        var _this = _super.call(this) || this;
        _this.name = name;
        return _this;
    }
    AliasMethodBlueprint.prototype.descriptor = function (target, key, meta) {
        return new AliasMethodDescriptor(this.name);
    };
    return AliasMethodBlueprint;
}(mixin_1.Blueprint));
function aliasMethod(name) {
    return new AliasMethodBlueprint(name);
}
exports.aliasMethod = aliasMethod;
var AliasBlueprint = (function (_super) {
    __extends(AliasBlueprint, _super);
    function AliasBlueprint(name) {
        var _this = this;
        var parent = name.slice(0, -1);
        var last = name[name.length - 1];
        var get = function () {
            return name.reduce(function (obj, n) { return obj[n]; }, this);
        };
        var set = function (value) {
            var p = parent.reduce(function (obj, n) { return obj[n]; }, this);
            p[last] = value;
        };
        _this = _super.call(this, { get: get, set: set }, [name]) || this;
        _this.name = name;
        return _this;
    }
    AliasBlueprint.prototype.descriptor = function (target, key, meta) {
        if (this.name[0] === key)
            throw new Error("Setting alias '" + key + "' on self");
        return _super.prototype.descriptor.call(this, target, key, meta);
    };
    return AliasBlueprint;
}(computed_1.ComputedBlueprint));
function alias(name) {
    return new AliasBlueprint(name.split('.'));
}
exports.alias = alias;
