// @glimmer/reference/lib/iterable.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var util_1 = require("@glimmer/util");
var ListItem = (function (_super) {
    __extends(ListItem, _super);
    function ListItem(iterable, result) {
        var _this = _super.call(this, iterable.valueReferenceFor(result)) || this;
        _this.retained = false;
        _this.seen = false;
        _this.key = result.key;
        _this.iterable = iterable;
        _this.memo = iterable.memoReferenceFor(result);
        return _this;
    }
    ListItem.prototype.update = function (item) {
        this.retained = true;
        this.iterable.updateValueReference(this.value, item);
        this.iterable.updateMemoReference(this.memo, item);
    };
    ListItem.prototype.shouldRemove = function () {
        return !this.retained;
    };
    ListItem.prototype.reset = function () {
        this.retained = false;
        this.seen = false;
    };
    return ListItem;
}(util_1.ListNode));
exports.ListItem = ListItem;
var IterationArtifacts = (function () {
    function IterationArtifacts(iterable) {
        this.map = util_1.dict();
        this.list = new util_1.LinkedList();
        this.tag = iterable.tag;
        this.iterable = iterable;
    }
    IterationArtifacts.prototype.isEmpty = function () {
        var iterator = this.iterator = this.iterable.iterate();
        return iterator.isEmpty();
    };
    IterationArtifacts.prototype.iterate = function () {
        var iterator = this.iterator || this.iterable.iterate();
        this.iterator = null;
        return iterator;
    };
    IterationArtifacts.prototype.has = function (key) {
        return !!this.map[key];
    };
    IterationArtifacts.prototype.get = function (key) {
        return this.map[key];
    };
    IterationArtifacts.prototype.wasSeen = function (key) {
        var node = this.map[key];
        return node && node.seen;
    };
    IterationArtifacts.prototype.append = function (item) {
        var _a = this, map = _a.map, list = _a.list, iterable = _a.iterable;
        var node = map[item.key] = new ListItem(iterable, item);
        list.append(node);
        return node;
    };
    IterationArtifacts.prototype.insertBefore = function (item, reference) {
        var _a = this, map = _a.map, list = _a.list, iterable = _a.iterable;
        var node = map[item.key] = new ListItem(iterable, item);
        node.retained = true;
        list.insertBefore(node, reference);
        return node;
    };
    IterationArtifacts.prototype.move = function (item, reference) {
        var list = this.list;
        item.retained = true;
        list.remove(item);
        list.insertBefore(item, reference);
    };
    IterationArtifacts.prototype.remove = function (item) {
        var list = this.list;
        list.remove(item);
        delete this.map[item.key];
    };
    IterationArtifacts.prototype.nextNode = function (item) {
        return this.list.nextNode(item);
    };
    IterationArtifacts.prototype.head = function () {
        return this.list.head();
    };
    return IterationArtifacts;
}());
exports.IterationArtifacts = IterationArtifacts;
var ReferenceIterator = (function () {
    // if anyone needs to construct this object with something other than
    // an iterable, let @wycats know.
    function ReferenceIterator(iterable) {
        this.iterator = null;
        var artifacts = new IterationArtifacts(iterable);
        this.artifacts = artifacts;
    }
    ReferenceIterator.prototype.next = function () {
        var artifacts = this.artifacts;
        var iterator = (this.iterator = this.iterator || artifacts.iterate());
        var item = iterator.next();
        if (!item)
            return null;
        return artifacts.append(item);
    };
    return ReferenceIterator;
}());
exports.ReferenceIterator = ReferenceIterator;
var Phase;
(function (Phase) {
    Phase[Phase["Append"] = 0] = "Append";
    Phase[Phase["Prune"] = 1] = "Prune";
    Phase[Phase["Done"] = 2] = "Done";
})(Phase || (Phase = {}));
var IteratorSynchronizer = (function () {
    function IteratorSynchronizer(_a) {
        var target = _a.target, artifacts = _a.artifacts;
        this.target = target;
        this.artifacts = artifacts;
        this.iterator = artifacts.iterate();
        this.current = artifacts.head();
    }
    IteratorSynchronizer.prototype.sync = function () {
        var phase = Phase.Append;
        while (true) {
            switch (phase) {
                case Phase.Append:
                    phase = this.nextAppend();
                    break;
                case Phase.Prune:
                    phase = this.nextPrune();
                    break;
                case Phase.Done:
                    this.nextDone();
                    return;
            }
        }
    };
    IteratorSynchronizer.prototype.advanceToKey = function (key) {
        var _a = this, current = _a.current, artifacts = _a.artifacts;
        var seek = current;
        while (seek && seek.key !== key) {
            seek.seen = true;
            seek = artifacts.nextNode(seek);
        }
        this.current = seek && artifacts.nextNode(seek);
    };
    IteratorSynchronizer.prototype.nextAppend = function () {
        var _a = this, iterator = _a.iterator, current = _a.current, artifacts = _a.artifacts;
        var item = iterator.next();
        if (item === null) {
            return this.startPrune();
        }
        var key = item.key;
        if (current && current.key === key) {
            this.nextRetain(item);
        }
        else if (artifacts.has(key)) {
            this.nextMove(item);
        }
        else {
            this.nextInsert(item);
        }
        return Phase.Append;
    };
    IteratorSynchronizer.prototype.nextRetain = function (item) {
        var _a = this, artifacts = _a.artifacts, current = _a.current;
        current = util_1.expect(current, 'BUG: current is empty');
        current.update(item);
        this.current = artifacts.nextNode(current);
        this.target.retain(item.key, current.value, current.memo);
    };
    IteratorSynchronizer.prototype.nextMove = function (item) {
        var _a = this, current = _a.current, artifacts = _a.artifacts, target = _a.target;
        var key = item.key;
        var found = artifacts.get(item.key);
        found.update(item);
        if (artifacts.wasSeen(item.key)) {
            artifacts.move(found, current);
            target.move(found.key, found.value, found.memo, current ? current.key : null);
        }
        else {
            this.advanceToKey(key);
        }
    };
    IteratorSynchronizer.prototype.nextInsert = function (item) {
        var _a = this, artifacts = _a.artifacts, target = _a.target, current = _a.current;
        var node = artifacts.insertBefore(item, current);
        target.insert(node.key, node.value, node.memo, current ? current.key : null);
    };
    IteratorSynchronizer.prototype.startPrune = function () {
        this.current = this.artifacts.head();
        return Phase.Prune;
    };
    IteratorSynchronizer.prototype.nextPrune = function () {
        var _a = this, artifacts = _a.artifacts, target = _a.target, current = _a.current;
        if (current === null) {
            return Phase.Done;
        }
        var node = current;
        this.current = artifacts.nextNode(node);
        if (node.shouldRemove()) {
            artifacts.remove(node);
            target.delete(node.key);
        }
        else {
            node.reset();
        }
        return Phase.Prune;
    };
    IteratorSynchronizer.prototype.nextDone = function () {
        this.target.done();
    };
    return IteratorSynchronizer;
}());
exports.IteratorSynchronizer = IteratorSynchronizer;
