// @glimmer/object-model/lib/blueprint.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var reference_1 = require("./reference");
var Blueprint = (function () {
    function Blueprint(key) {
        this.key = key;
    }
    return Blueprint;
}());
exports.Blueprint = Blueprint;
var VALUE_DESCRIPTOR = {
    enumerable: true,
    configurable: true,
    writable: true,
    value: undefined
};
var ACCESSOR_DESCRIPTOR = {
    enumerable: true,
    configurable: true,
    get: undefined,
    set: undefined
};
var ValueBlueprint = (function (_super) {
    __extends(ValueBlueprint, _super);
    function ValueBlueprint(key, value) {
        var _this = _super.call(this, key) || this;
        _this.value = value;
        return _this;
    }
    ValueBlueprint.prototype.define = function (home) {
        VALUE_DESCRIPTOR.value = this.value;
        Object.defineProperty(home, this.key, VALUE_DESCRIPTOR);
    };
    return ValueBlueprint;
}(Blueprint));
exports.ValueBlueprint = ValueBlueprint;
var GlimmerDescriptor = (function () {
    function GlimmerDescriptor() {
    }
    return GlimmerDescriptor;
}());
exports.GlimmerDescriptor = GlimmerDescriptor;
var Computed = (function (_super) {
    __extends(Computed, _super);
    function Computed(dependentKeys, accessor) {
        var _this = _super.call(this) || this;
        _this.dependentKeys = dependentKeys;
        _this.accessor = accessor;
        return _this;
    }
    Computed.prototype.blueprint = function (key) {
        return new ComputedBlueprint(key, this);
    };
    Computed.prototype.reference = function (root, key) {
    };
    Computed.prototype.define = function (home, key) {
        reference_1.classMeta(home).defineComputed(key, this);
        ACCESSOR_DESCRIPTOR.get = this.accessor.get;
        ACCESSOR_DESCRIPTOR.set = this.accessor.set;
        Object.defineProperty(home, key, ACCESSOR_DESCRIPTOR);
    };
    return Computed;
}(GlimmerDescriptor));
exports.Computed = Computed;
function computed() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    var depKeys = args.slice(0, -1);
    var accessor = args[args.length - 1];
    return new Computed(depKeys, accessor);
}
exports.computed = computed;
var ComputedBlueprint = (function (_super) {
    __extends(ComputedBlueprint, _super);
    function ComputedBlueprint(key, computed) {
        var _this = _super.call(this, key) || this;
        _this.computed = computed;
        return _this;
    }
    ComputedBlueprint.prototype.define = function (home) {
        this.computed.define(home, this.key);
    };
    return ComputedBlueprint;
}(Blueprint));
exports.ComputedBlueprint = ComputedBlueprint;
var DescriptorBlueprint = (function (_super) {
    __extends(DescriptorBlueprint, _super);
    function DescriptorBlueprint(key, descriptor) {
        var _this = _super.call(this, key) || this;
        _this.descriptor = descriptor;
        return _this;
    }
    DescriptorBlueprint.prototype.define = function (home) {
        Object.defineProperty(home, this.key, this.descriptor);
    };
    return DescriptorBlueprint;
}(Blueprint));
exports.DescriptorBlueprint = DescriptorBlueprint;
