// @glimmer/object-model/lib/meta.js
"use strict";
var util_1 = require("@glimmer/util");
var reference_1 = require("@glimmer/reference");
var default_1 = (function () {
    function default_1() {
        this.bookkeeping = util_1.dict();
    }
    default_1.prototype.tag = function (name) {
        var bookkeeping = this.bookkeeping;
        var tag = bookkeeping[name];
        if (tag === undefined) {
            tag = new reference_1.DirtyableTag();
            bookkeeping[name] = tag;
        }
        return tag;
    };
    default_1.prototype.dirty = function (name) {
        var tag = this.tag(name);
        tag.dirty();
    };
    return default_1;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
var ClassMeta = (function () {
    function ClassMeta() {
        this.computed = util_1.dict();
    }
    ClassMeta.prototype.defineComputed = function (name, value) {
        this.computed[name] = value;
    };
    ClassMeta.prototype.getComputed = function (name) {
        return this.computed[name];
    };
    return ClassMeta;
}());
exports.ClassMeta = ClassMeta;
