// @glimmer/object-model/lib/object.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var blueprint_1 = require("./blueprint");
var GlimmerObjectBase = (function () {
    function GlimmerObjectBase() {
    }
    return GlimmerObjectBase;
}());
exports.GlimmerObjectBase = GlimmerObjectBase;
var GlimmerObject = (function () {
    function GlimmerObject(properties) {
        if (properties !== undefined) {
            Object.assign(this, properties);
        }
    }
    GlimmerObject.create = function (properties) {
        return new this(properties);
    };
    GlimmerObject.extend = function (extensions) {
        // This method intentionally uses internal typecasts to convince TypeScript
        // that what we're doing is legitimate. The real action of this method is in
        // its signature (and the signature of GlimmerClass).
        var sub = (function (_super) {
            __extends(class_1, _super);
            function class_1() {
                return _super !== null && _super.apply(this, arguments) || this;
            }
            return class_1;
        }(this));
        if (extensions !== undefined) {
            var blueprints = Object.keys(extensions).map(function (k) {
                var extension = extensions[k];
                if (extension instanceof blueprint_1.GlimmerDescriptor) {
                    return extension.blueprint(k);
                }
                else {
                    return new blueprint_1.ValueBlueprint(k, extensions[k]);
                }
            });
            blueprints.forEach(function (b) { return b.define(sub.prototype); });
        }
        return sub;
    };
    return GlimmerObject;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = GlimmerObject;
/**
 * Adapter for TypeScript.
 *
 * A shim library could just declare the variable as the type
 * GlimmerClass<FirstInstance>. The key point is that shimming
 * between the two worlds requires writing a static interface
 * definition, as TypeScript isn't willing to treat the dynamic
 * intersection as a valid constructor return type.
 *
 * This function is a noop and is used like this:
 *
 * ```ts
 * // in one file
 * let Person = GlimmerObject.extend({
 *   named: 'Dan'
 * });
 *
 *
 * // in a TypeScript wrapper file
 * import { Person as OriginalPerson } from 'original-definition';
 *
 * interface PersonInstance {
 *   named: string
 * }
 *
 * export let Person = classof<PersonInstance>(OriginalPerson);
 *
 * // in a subclassing file
 * import { Person } from 'wrapped-definition';
 *
 * class FancyPerson extends Person {
 *   salutation: string;
 *
 *   fullName() {
 *     return `${this.salutation} ${this.named}`;
 *   }
 * }
 *
 * FancyPerson.create({ name: 'Tom Dale', salutation: 'Mr.' }).fullName();
 * // typechecks and returns 'Mr. Tom Dale'
 * ```
*/
function classof(klass) {
    return klass;
}
exports.classof = classof;
