// glimmer-engine/packages/@glimmer/runtime/dist/commonjs/lib/compiled/expressions/concat.js
"use strict";
const reference_1 = require("@glimmer/reference");
class CompiledConcat {
    constructor(parts) {
        this.parts = parts;
        this.type = "concat";
    }
    evaluate(vm) {
        let parts = new Array(this.parts.length);
        for (let i = 0; i < this.parts.length; i++) {
            parts[i] = this.parts[i].evaluate(vm);
        }
        return new ConcatReference(parts);
    }
    toJSON() {
        return `concat(${this.parts.map(expr => expr.toJSON()).join(", ")})`;
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledConcat;
class ConcatReference extends reference_1.CachedReference {
    constructor(parts) {
        super();
        this.parts = parts;
        this.tag = reference_1.combineTagged(parts);
    }
    compute() {
        let parts = new Array();
        for (let i = 0; i < this.parts.length; i++) {
            let value = this.parts[i].value();
            if (value !== null && value !== undefined) {
                parts[i] = castToString(value);
            }
        }
        if (parts.length > 0) {
            return parts.join('');
        }
        return null;
    }
}
exports.ConcatReference = ConcatReference;
function castToString(value) {
    if (typeof value['toString'] !== 'function') {
        return '';
    }
    return String(value);
}
