// glimmer-engine/packages/@glimmer/runtime/dist/commonjs/lib/compiled/expressions/named-args.js
"use strict";
const references_1 = require("../../references");
const utils_1 = require("../../utils");
const reference_1 = require("@glimmer/reference");
const util_1 = require("@glimmer/util");
class CompiledNamedArgs {
    constructor(keys, values) {
        this.keys = keys;
        this.values = values;
        this.length = keys.length;
        util_1.assert(keys.length === values.length, 'Keys and values do not have the same length');
    }
    static empty() {
        return exports.COMPILED_EMPTY_NAMED_ARGS;
    }
    static create(map) {
        let keys = Object.keys(map);
        let length = keys.length;
        if (length > 0) {
            let values = [];
            for (let i = 0; i < length; i++) {
                values[i] = map[keys[i]];
            }
            return new this(keys, values);
        }
        else {
            return exports.COMPILED_EMPTY_NAMED_ARGS;
        }
    }
    evaluate(vm) {
        let { keys, values, length } = this;
        let evaluated = new Array(length);
        for (let i = 0; i < length; i++) {
            evaluated[i] = values[i].evaluate(vm);
        }
        return new EvaluatedNamedArgs(keys, evaluated);
    }
    toJSON() {
        let { keys, values } = this;
        let inner = keys.map((key, i) => `${key}: ${values[i].toJSON()}`).join(", ");
        return `{${inner}}`;
    }
}
exports.CompiledNamedArgs = CompiledNamedArgs;
exports.COMPILED_EMPTY_NAMED_ARGS = new (class extends CompiledNamedArgs {
    constructor() {
        super(utils_1.EMPTY_ARRAY, utils_1.EMPTY_ARRAY);
    }
    evaluate(_vm) {
        return exports.EVALUATED_EMPTY_NAMED_ARGS;
    }
    toJSON() {
        return `<EMPTY>`;
    }
});
class EvaluatedNamedArgs {
    constructor(keys, values, _map = null) {
        this.keys = keys;
        this.values = values;
        this._map = _map;
        this.tag = reference_1.combineTagged(values);
        this.length = keys.length;
        util_1.assert(keys.length === values.length, 'Keys and values do not have the same length');
    }
    static create(map) {
        let keys = Object.keys(map);
        let length = keys.length;
        if (length > 0) {
            let values = new Array(length);
            for (let i = 0; i < length; i++) {
                values[i] = map[keys[i]];
            }
            return new this(keys, values, map);
        }
        else {
            return exports.EVALUATED_EMPTY_NAMED_ARGS;
        }
    }
    static empty() {
        return exports.EVALUATED_EMPTY_NAMED_ARGS;
    }
    get map() {
        let { _map: map } = this;
        if (map) {
            return map;
        }
        map = this._map = util_1.dict();
        let { keys, values, length } = this;
        for (let i = 0; i < length; i++) {
            map[keys[i]] = values[i];
        }
        return map;
    }
    get(key) {
        let { keys, values } = this;
        let index = keys.indexOf(key);
        return (index === -1) ? references_1.UNDEFINED_REFERENCE : values[index];
    }
    has(key) {
        return this.keys.indexOf(key) !== -1;
    }
    value() {
        let { keys, values } = this;
        let out = util_1.dict();
        for (let i = 0; i < keys.length; i++) {
            let key = keys[i];
            let ref = values[i];
            out[key] = ref.value();
        }
        return out;
    }
}
exports.EvaluatedNamedArgs = EvaluatedNamedArgs;
exports.EVALUATED_EMPTY_NAMED_ARGS = new (class extends EvaluatedNamedArgs {
    constructor() {
        super(utils_1.EMPTY_ARRAY, utils_1.EMPTY_ARRAY, utils_1.EMPTY_DICT);
    }
    get() {
        return references_1.UNDEFINED_REFERENCE;
    }
    has(_key) {
        return false;
    }
    value() {
        return utils_1.EMPTY_DICT;
    }
});
