// glimmer-engine/dist/es6/@glimmer/runtime/lib/opcodes.js
import { LOGGER, initializeGuid, fillNulls } from '@glimmer/util';
import { NULL_REFERENCE, UNDEFINED_REFERENCE } from './references';
export function pretty(json) {
    return `${json.type}(${json.args ? json.args.join(', ') : ''})`;
}
export function defaultToJSON(opcode) {
    return { type: opcode[0] };
}
export var OpcodeName;
(function (OpcodeName) {
    OpcodeName[OpcodeName["PushChildScope"] = 0] = "PushChildScope";
    OpcodeName[OpcodeName["PopScope"] = 1] = "PopScope";
    OpcodeName[OpcodeName["PushDynamicScope"] = 2] = "PushDynamicScope";
    OpcodeName[OpcodeName["PopDynamicScope"] = 3] = "PopDynamicScope";
    OpcodeName[OpcodeName["Put"] = 4] = "Put";
    OpcodeName[OpcodeName["EvaluatePut"] = 5] = "EvaluatePut";
    OpcodeName[OpcodeName["PutArgs"] = 6] = "PutArgs";
    OpcodeName[OpcodeName["BindPositionalArgs"] = 7] = "BindPositionalArgs";
    OpcodeName[OpcodeName["BindNamedArgs"] = 8] = "BindNamedArgs";
    OpcodeName[OpcodeName["BindBlocks"] = 9] = "BindBlocks";
    OpcodeName[OpcodeName["BindPartialArgs"] = 10] = "BindPartialArgs";
    OpcodeName[OpcodeName["BindCallerScope"] = 11] = "BindCallerScope";
    OpcodeName[OpcodeName["BindDynamicScope"] = 12] = "BindDynamicScope";
    OpcodeName[OpcodeName["Enter"] = 13] = "Enter";
    OpcodeName[OpcodeName["Exit"] = 14] = "Exit";
    OpcodeName[OpcodeName["Evaluate"] = 15] = "Evaluate";
    OpcodeName[OpcodeName["Jump"] = 16] = "Jump";
    OpcodeName[OpcodeName["JumpIf"] = 17] = "JumpIf";
    OpcodeName[OpcodeName["JumpUnless"] = 18] = "JumpUnless";
    OpcodeName[OpcodeName["Test"] = 19] = "Test";
    OpcodeName[OpcodeName["OpenBlock"] = 20] = "OpenBlock";
    OpcodeName[OpcodeName["CloseBlock"] = 21] = "CloseBlock";
    OpcodeName[OpcodeName["PutDynamicComponent"] = 22] = "PutDynamicComponent";
    OpcodeName[OpcodeName["PutComponent"] = 23] = "PutComponent";
    OpcodeName[OpcodeName["OpenComponent"] = 24] = "OpenComponent";
    OpcodeName[OpcodeName["DidCreateElement"] = 25] = "DidCreateElement";
    OpcodeName[OpcodeName["ShadowAttributes"] = 26] = "ShadowAttributes";
    OpcodeName[OpcodeName["DidRenderLayout"] = 27] = "DidRenderLayout";
    OpcodeName[OpcodeName["CloseComponent"] = 28] = "CloseComponent";
    OpcodeName[OpcodeName["Text"] = 29] = "Text";
    OpcodeName[OpcodeName["Comment"] = 30] = "Comment";
    OpcodeName[OpcodeName["DynamicContent"] = 31] = "DynamicContent";
    OpcodeName[OpcodeName["OpenElement"] = 32] = "OpenElement";
    OpcodeName[OpcodeName["PushRemoteElement"] = 33] = "PushRemoteElement";
    OpcodeName[OpcodeName["PopRemoteElement"] = 34] = "PopRemoteElement";
    OpcodeName[OpcodeName["OpenComponentElement"] = 35] = "OpenComponentElement";
    OpcodeName[OpcodeName["OpenDynamicElement"] = 36] = "OpenDynamicElement";
    OpcodeName[OpcodeName["FlushElement"] = 37] = "FlushElement";
    OpcodeName[OpcodeName["CloseElement"] = 38] = "CloseElement";
    OpcodeName[OpcodeName["PopElement"] = 39] = "PopElement";
    OpcodeName[OpcodeName["StaticAttr"] = 40] = "StaticAttr";
    OpcodeName[OpcodeName["Modifier"] = 41] = "Modifier";
    OpcodeName[OpcodeName["DynamicAttrNS"] = 42] = "DynamicAttrNS";
    OpcodeName[OpcodeName["DynamicAttr"] = 43] = "DynamicAttr";
    OpcodeName[OpcodeName["PutIterator"] = 44] = "PutIterator";
    OpcodeName[OpcodeName["EnterList"] = 45] = "EnterList";
    OpcodeName[OpcodeName["ExitList"] = 46] = "ExitList";
    OpcodeName[OpcodeName["EnterWithKey"] = 47] = "EnterWithKey";
    OpcodeName[OpcodeName["NextIter"] = 48] = "NextIter";
    OpcodeName[OpcodeName["PutDynamicPartial"] = 49] = "PutDynamicPartial";
    OpcodeName[OpcodeName["PutPartial"] = 50] = "PutPartial";
    OpcodeName[OpcodeName["EvaluatePartial"] = 51] = "EvaluatePartial"; // (Other<SymbolTable>, Other<Dict<PartialBlock>>)
})(OpcodeName || (OpcodeName = {}));
export class Constants {
    constructor() {
        // `0` means NULL
        this.references = [];
        this.strings = [];
        this.expressions = [];
        this.arrays = [];
        this.slices = [];
        this.blocks = [];
        this.functions = [];
        this.others = [];
        this.NULL_REFERENCE = this.reference(NULL_REFERENCE);
        this.UNDEFINED_REFERENCE = this.reference(UNDEFINED_REFERENCE);
    }
    getReference(value) {
        return this.references[value - 1];
    }
    reference(value) {
        let index = this.references.length;
        this.references.push(value);
        return index + 1;
    }
    getString(value) {
        return this.strings[value - 1];
    }
    string(value) {
        let index = this.strings.length;
        this.strings.push(value);
        return index + 1;
    }
    getExpression(value) {
        return this.expressions[value - 1];
    }
    expression(value) {
        let index = this.expressions.length;
        this.expressions.push(value);
        return index + 1;
    }
    getArray(value) {
        return this.arrays[value - 1];
    }
    array(values) {
        let index = this.arrays.length;
        this.arrays.push(values);
        return index + 1;
    }
    getSlice(value) {
        return this.slices[value - 1];
    }
    slice(slice) {
        // TODO: Put the entire program in one big array
        let index = this.slices.length;
        this.slices.push(slice);
        return index + 1;
    }
    getBlock(value) {
        return this.blocks[value - 1];
    }
    block(block) {
        let index = this.blocks.length;
        this.blocks.push(block);
        return index + 1;
    }
    getFunction(value) {
        return this.functions[value - 1];
    }
    function(f) {
        let index = this.functions.length;
        this.functions.push(f);
        return index + 1;
    }
    getOther(value) {
        return this.others[value - 1];
    }
    other(other) {
        let index = this.others.length;
        this.others.push(other);
        return index + 1;
    }
}
export class AppendOpcodes {
    constructor() {
        this.evaluateOpcode = fillNulls(OpcodeName.EvaluatePartial + 1);
    }
    add(name, evaluate) {
        this.evaluateOpcode[name] = evaluate;
    }
    construct(name, _debug, op1, op2, op3) {
        return [name | 0, (op1 || 0) | 0, (op2 || 0) | 0, (op3 || 0) | 0];
    }
    evaluate(vm, opcode) {
        LOGGER.debug(`[VM] OPCODE: ${opcode.type}`);
        let func = this.evaluateOpcode[opcode.type];
        func(vm, opcode);
    }
}
export const APPEND_OPCODES = new AppendOpcodes();
export class AbstractOpcode {
    constructor() {
        initializeGuid(this);
    }
    toJSON() {
        return { guid: this._guid, type: this.type };
    }
}
export class UpdatingOpcode extends AbstractOpcode {
    constructor() {
        super(...arguments);
        this.next = null;
        this.prev = null;
    }
}
export function inspect(opcodes) {
    let buffer = [];
    opcodes.forEach((opcode, i) => {
        _inspect(opcode.toJSON(), buffer, 0, i);
    });
    return buffer.join('');
}
function _inspect(opcode, buffer, level, index) {
    let indentation = [];
    for (let i = 0; i < level; i++) {
        indentation.push('  ');
    }
    buffer.push(...indentation);
    buffer.push(`${index}. ${opcode.type}`);
    if (opcode.args || opcode.details) {
        buffer.push('(');
        if (opcode.args) {
            buffer.push(opcode.args.join(', '));
        }
        if (opcode.details) {
            let keys = Object.keys(opcode.details);
            if (keys.length) {
                if (opcode.args && opcode.args.length) {
                    buffer.push(', ');
                }
                buffer.push(keys.map(key => `${key}=${opcode.details && opcode.details[key]}`).join(', '));
            }
        }
        buffer.push(')');
    }
    buffer.push('\n');
    if (opcode.children && opcode.children.length) {
        for (let i = 0; i < opcode.children.length; i++) {
            _inspect(opcode.children[i], buffer, level + 1, i);
        }
    }
}
