// glimmer-engine/dist/named-amd/@glimmer/object/lib/descriptors.js
enifed('@glimmer/object/lib/descriptors', ['exports', '@glimmer/object/lib/mixin', '@glimmer/object/lib/computed'], function (exports, _glimmerObjectLibMixin, _glimmerObjectLibComputed) {
    'use strict';

    exports.aliasMethod = aliasMethod;
    exports.alias = alias;

    function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

    var AliasMethodDescriptor = (function (_Descriptor) {
        _inherits(AliasMethodDescriptor, _Descriptor);

        function AliasMethodDescriptor(name) {
            _classCallCheck(this, AliasMethodDescriptor);

            _Descriptor.call(this);
            this.name = name;
        }

        AliasMethodDescriptor.prototype.define = function define(target, key, home) {
            var name = this.name;
            Object.defineProperty(target, key, {
                enumerable: true,
                configurable: true,
                get: function () {
                    return this[name];
                }
            });
        };

        return AliasMethodDescriptor;
    })(_glimmerObjectLibMixin.Descriptor);

    var AliasMethodBlueprint = (function (_Blueprint) {
        _inherits(AliasMethodBlueprint, _Blueprint);

        function AliasMethodBlueprint(name) {
            _classCallCheck(this, AliasMethodBlueprint);

            _Blueprint.call(this);
            this.name = name;
        }

        AliasMethodBlueprint.prototype.descriptor = function descriptor(target, key, meta) {
            return new AliasMethodDescriptor(this.name);
        };

        return AliasMethodBlueprint;
    })(_glimmerObjectLibMixin.Blueprint);

    function aliasMethod(name) {
        return new AliasMethodBlueprint(name);
    }

    var AliasBlueprint = (function (_ComputedBlueprint) {
        _inherits(AliasBlueprint, _ComputedBlueprint);

        function AliasBlueprint(name) {
            _classCallCheck(this, AliasBlueprint);

            var parent = name.slice(0, -1);
            var last = name[name.length - 1];
            var get = function () {
                return name.reduce(function (obj, n) {
                    return obj[n];
                }, this);
            };
            var set = function (value) {
                var p = parent.reduce(function (obj, n) {
                    return obj[n];
                }, this);
                p[last] = value;
            };
            _ComputedBlueprint.call(this, { get: get, set: set }, [name]);
            this.name = name;
        }

        AliasBlueprint.prototype.descriptor = function descriptor(target, key, meta) {
            if (this.name[0] === key) throw new Error('Setting alias \'' + key + '\' on self');
            return _ComputedBlueprint.prototype.descriptor.call(this, target, key, meta);
        };

        return AliasBlueprint;
    })(_glimmerObjectLibComputed.ComputedBlueprint);

    function alias(name) {
        return new AliasBlueprint(name.split('.'));
    }
});
