// glimmer-engine/dist/named-amd/@glimmer/object-model/lib/blueprint.js
enifed('@glimmer/object-model/lib/blueprint', ['exports', '@glimmer/object-model/lib/reference'], function (exports, _glimmerObjectModelLibReference) {
    'use strict';

    exports.computed = computed;

    function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

    function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var Blueprint = function Blueprint(key) {
        _classCallCheck(this, Blueprint);

        this.key = key;
    };

    exports.Blueprint = Blueprint;

    var VALUE_DESCRIPTOR = {
        enumerable: true,
        configurable: true,
        writable: true,
        value: undefined
    };
    var ACCESSOR_DESCRIPTOR = {
        enumerable: true,
        configurable: true,
        get: undefined,
        set: undefined
    };

    var ValueBlueprint = (function (_Blueprint) {
        _inherits(ValueBlueprint, _Blueprint);

        function ValueBlueprint(key, value) {
            _classCallCheck(this, ValueBlueprint);

            _Blueprint.call(this, key);
            this.value = value;
        }

        ValueBlueprint.prototype.define = function define(home) {
            VALUE_DESCRIPTOR.value = this.value;
            Object.defineProperty(home, this.key, VALUE_DESCRIPTOR);
        };

        return ValueBlueprint;
    })(Blueprint);

    exports.ValueBlueprint = ValueBlueprint;

    var GlimmerDescriptor = function GlimmerDescriptor() {
        _classCallCheck(this, GlimmerDescriptor);
    };

    exports.GlimmerDescriptor = GlimmerDescriptor;

    var Computed = (function (_GlimmerDescriptor) {
        _inherits(Computed, _GlimmerDescriptor);

        function Computed(dependentKeys, accessor) {
            _classCallCheck(this, Computed);

            _GlimmerDescriptor.call(this);
            this.dependentKeys = dependentKeys;
            this.accessor = accessor;
        }

        Computed.prototype.blueprint = function blueprint(key) {
            return new ComputedBlueprint(key, this);
        };

        Computed.prototype.reference = function reference(root, key) {};

        Computed.prototype.define = function define(home, key) {
            _glimmerObjectModelLibReference.classMeta(home).defineComputed(key, this);
            ACCESSOR_DESCRIPTOR.get = this.accessor.get;
            ACCESSOR_DESCRIPTOR.set = this.accessor.set;
            Object.defineProperty(home, key, ACCESSOR_DESCRIPTOR);
        };

        return Computed;
    })(GlimmerDescriptor);

    exports.Computed = Computed;

    function computed() {
        for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
            args[_key] = arguments[_key];
        }

        var depKeys = args.slice(0, -1);
        var accessor = args[args.length - 1];
        return new Computed(depKeys, accessor);
    }

    var ComputedBlueprint = (function (_Blueprint2) {
        _inherits(ComputedBlueprint, _Blueprint2);

        function ComputedBlueprint(key, computed) {
            _classCallCheck(this, ComputedBlueprint);

            _Blueprint2.call(this, key);
            this.computed = computed;
        }

        ComputedBlueprint.prototype.define = function define(home) {
            this.computed.define(home, this.key);
        };

        return ComputedBlueprint;
    })(Blueprint);

    exports.ComputedBlueprint = ComputedBlueprint;

    var DescriptorBlueprint = (function (_Blueprint3) {
        _inherits(DescriptorBlueprint, _Blueprint3);

        function DescriptorBlueprint(key, descriptor) {
            _classCallCheck(this, DescriptorBlueprint);

            _Blueprint3.call(this, key);
            this.descriptor = descriptor;
        }

        DescriptorBlueprint.prototype.define = function define(home) {
            Object.defineProperty(home, this.key, this.descriptor);
        };

        return DescriptorBlueprint;
    })(Blueprint);

    exports.DescriptorBlueprint = DescriptorBlueprint;
});
