// glimmer-engine/dist/named-amd/@glimmer/object-model/lib/meta.js
enifed('@glimmer/object-model/lib/meta', ['exports', '@glimmer/util', '@glimmer/reference'], function (exports, _glimmerUtil, _glimmerReference) {
    'use strict';

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var _default = (function () {
        function _default() {
            _classCallCheck(this, _default);

            this.bookkeeping = _glimmerUtil.dict();
        }

        _default.prototype.tag = function tag(name) {
            var bookkeeping = this.bookkeeping;
            var tag = bookkeeping[name];
            if (tag === undefined) {
                tag = new _glimmerReference.DirtyableTag();
                bookkeeping[name] = tag;
            }
            return tag;
        };

        _default.prototype.dirty = function dirty(name) {
            var tag = this.tag(name);
            tag.dirty();
        };

        return _default;
    })();

    exports.default = _default;

    var ClassMeta = (function () {
        function ClassMeta() {
            _classCallCheck(this, ClassMeta);

            this.computed = _glimmerUtil.dict();
        }

        ClassMeta.prototype.defineComputed = function defineComputed(name, value) {
            this.computed[name] = value;
        };

        ClassMeta.prototype.getComputed = function getComputed(name) {
            return this.computed[name];
        };

        return ClassMeta;
    })();

    exports.ClassMeta = ClassMeta;
});
