// glimmer-engine/dist/named-amd/@glimmer/object-model/lib/object.js
enifed('@glimmer/object-model/lib/object', ['exports', '@glimmer/object-model/lib/blueprint'], function (exports, _glimmerObjectModelLibBlueprint) {
    'use strict';

    exports.classof = classof;

    function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

    function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var GlimmerObjectBase = function GlimmerObjectBase() {
        _classCallCheck(this, GlimmerObjectBase);
    };

    exports.GlimmerObjectBase = GlimmerObjectBase;

    var GlimmerObject = (function () {
        GlimmerObject.create = function create(properties) {
            return new this(properties);
        };

        GlimmerObject.extend = function extend(extensions) {
            // This method intentionally uses internal typecasts to convince TypeScript
            // that what we're doing is legitimate. The real action of this method is in
            // its signature (and the signature of GlimmerClass).
            var sub = (function (_ref) {
                _inherits(sub, _ref);

                function sub() {
                    _classCallCheck(this, sub);

                    _ref.apply(this, arguments);
                }

                return sub;
            })(this);
            if (extensions !== undefined) {
                var blueprints = Object.keys(extensions).map(function (k) {
                    var extension = extensions[k];
                    if (extension instanceof _glimmerObjectModelLibBlueprint.GlimmerDescriptor) {
                        return extension.blueprint(k);
                    } else {
                        return new _glimmerObjectModelLibBlueprint.ValueBlueprint(k, extensions[k]);
                    }
                });
                blueprints.forEach(function (b) {
                    return b.define(sub.prototype);
                });
            }
            return sub;
        };

        function GlimmerObject(properties) {
            _classCallCheck(this, GlimmerObject);

            if (properties !== undefined) {
                Object.assign(this, properties);
            }
        }

        /**
         * Adapter for TypeScript.
         *
         * A shim library could just declare the variable as the type
         * GlimmerClass<FirstInstance>. The key point is that shimming
         * between the two worlds requires writing a static interface
         * definition, as TypeScript isn't willing to treat the dynamic
         * intersection as a valid constructor return type.
         *
         * This function is a noop and is used like this:
         *
         * ```ts
         * // in one file
         * let Person = GlimmerObject.extend({
         *   named: 'Dan'
         * });
         *
         *
         * // in a TypeScript wrapper file
         * import { Person as OriginalPerson } from 'original-definition';
         *
         * interface PersonInstance {
         *   named: string
         * }
         *
         * export let Person = classof<PersonInstance>(OriginalPerson);
         *
         * // in a subclassing file
         * import { Person } from 'wrapped-definition';
         *
         * class FancyPerson extends Person {
         *   salutation: string;
         *
         *   fullName() {
         *     return `${this.salutation} ${this.named}`;
         *   }
         * }
         *
         * FancyPerson.create({ name: 'Tom Dale', salutation: 'Mr.' }).fullName();
         * // typechecks and returns 'Mr. Tom Dale'
         * ```
        */
        return GlimmerObject;
    })();

    exports.default = GlimmerObject;

    function classof(klass) {
        return klass;
    }
});
