// glimmer-engine/dist/named-amd/@glimmer/runtime/lib/vm/frame.js
enifed('@glimmer/runtime/lib/vm/frame', ['exports', '@glimmer/util'], function (exports, _glimmerUtil) {
    'use strict';

    var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var CapturedFrame = function CapturedFrame(operand, args, condition) {
        _classCallCheck(this, CapturedFrame);

        this.operand = operand;
        this.args = args;
        this.condition = condition;
    };

    exports.CapturedFrame = CapturedFrame;

    var Frame = (function () {
        function Frame(ops) {
            var component = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];
            var manager = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];
            var shadow = arguments.length <= 3 || arguments[3] === undefined ? null : arguments[3];

            _classCallCheck(this, Frame);

            this.ops = ops;
            this.component = component;
            this.manager = manager;
            this.shadow = shadow;
            this.operand = null;
            this.immediate = null;
            this.args = null;
            this.callerScope = null;
            this.blocks = null;
            this.condition = null;
            this.iterator = null;
            this.key = null;
            this.ip = ops[0];
        }

        Frame.prototype.capture = function capture() {
            return new CapturedFrame(this.operand, this.args, this.condition);
        };

        Frame.prototype.restore = function restore(frame) {
            this.operand = frame['operand'];
            this.args = frame['args'];
            this.condition = frame['condition'];
        };

        return Frame;
    })();

    var FrameStack = (function () {
        function FrameStack() {
            _classCallCheck(this, FrameStack);

            this.frames = [];
            this.frame = null;
        }

        FrameStack.prototype.push = function push(ops) {
            var component = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];
            var manager = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];
            var shadow = arguments.length <= 3 || arguments[3] === undefined ? null : arguments[3];

            var frame = this.frame === null ? this.frame = 0 : ++this.frame;
            if (this.frames.length <= frame) {
                this.frames.push(null);
            }
            this.frames[frame] = new Frame(ops, component, manager, shadow);
        };

        FrameStack.prototype.pop = function pop() {
            var frames = this.frames;
            var frame = this.frame;

            frames[_glimmerUtil.expect(frame, 'only pop after pushing')] = null;
            this.frame = frame === 0 ? null : frame - 1;
        };

        FrameStack.prototype.capture = function capture() {
            return this.currentFrame.capture();
        };

        FrameStack.prototype.restore = function restore(frame) {
            this.currentFrame.restore(frame);
        };

        FrameStack.prototype.getOps = function getOps() {
            return this.currentFrame.ops;
        };

        FrameStack.prototype.getCurrent = function getCurrent() {
            return this.currentFrame.ip;
        };

        FrameStack.prototype.setCurrent = function setCurrent(ip) {
            return this.currentFrame.ip = ip;
        };

        FrameStack.prototype.getOperand = function getOperand() {
            return _glimmerUtil.unwrap(this.currentFrame.operand);
        };

        FrameStack.prototype.setOperand = function setOperand(operand) {
            return this.currentFrame.operand = operand;
        };

        FrameStack.prototype.getImmediate = function getImmediate() {
            return this.currentFrame.immediate;
        };

        FrameStack.prototype.setImmediate = function setImmediate(value) {
            return this.currentFrame.immediate = value;
        };

        // FIXME: These options are required in practice by the existing code, but
        // figure out why.

        FrameStack.prototype.getArgs = function getArgs() {
            return this.currentFrame.args;
        };

        FrameStack.prototype.setArgs = function setArgs(args) {
            return this.currentFrame.args = args;
        };

        FrameStack.prototype.getCondition = function getCondition() {
            return _glimmerUtil.unwrap(this.currentFrame.condition);
        };

        FrameStack.prototype.setCondition = function setCondition(condition) {
            return this.currentFrame.condition = condition;
        };

        FrameStack.prototype.getIterator = function getIterator() {
            return _glimmerUtil.unwrap(this.currentFrame.iterator);
        };

        FrameStack.prototype.setIterator = function setIterator(iterator) {
            return this.currentFrame.iterator = iterator;
        };

        FrameStack.prototype.getKey = function getKey() {
            return this.currentFrame.key;
        };

        FrameStack.prototype.setKey = function setKey(key) {
            return this.currentFrame.key = key;
        };

        FrameStack.prototype.getBlocks = function getBlocks() {
            return _glimmerUtil.unwrap(this.currentFrame.blocks);
        };

        FrameStack.prototype.setBlocks = function setBlocks(blocks) {
            return this.currentFrame.blocks = blocks;
        };

        FrameStack.prototype.getCallerScope = function getCallerScope() {
            return _glimmerUtil.unwrap(this.currentFrame.callerScope);
        };

        FrameStack.prototype.setCallerScope = function setCallerScope(callerScope) {
            return this.currentFrame.callerScope = callerScope;
        };

        FrameStack.prototype.getComponent = function getComponent() {
            return _glimmerUtil.unwrap(this.currentFrame.component);
        };

        FrameStack.prototype.getManager = function getManager() {
            return _glimmerUtil.unwrap(this.currentFrame.manager);
        };

        FrameStack.prototype.getShadow = function getShadow() {
            return this.currentFrame.shadow;
        };

        FrameStack.prototype.goto = function goto(ip) {
            this.setCurrent(ip);
        };

        FrameStack.prototype.hasOpcodes = function hasOpcodes() {
            return this.frame !== null;
        };

        FrameStack.prototype.nextStatement = function nextStatement(env) {
            var ip = this.frames[_glimmerUtil.unwrap(this.frame)].ip;
            var ops = this.getOps();
            if (ip <= ops[1]) {
                var program = env.program;
                this.setCurrent(ip + 4);
                return program.opcode(ip);
            } else {
                this.pop();
                return null;
            }
        };

        _createClass(FrameStack, [{
            key: 'currentFrame',
            get: function () {
                return this.frames[_glimmerUtil.unwrap(this.frame)];
            }
        }]);

        return FrameStack;
    })();

    exports.FrameStack = FrameStack;
});
