// @glimmer/util/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var namespaces_1 = require("./lib/namespaces");
exports.getAttrNamespace = namespaces_1.getAttrNamespace;
__export(require("./lib/platform-utils"));
var assert_1 = require("./lib/assert");
exports.assert = assert_1.default;
var logger_1 = require("./lib/logger");
exports.LOGGER = logger_1.default;
exports.Logger = logger_1.Logger;
exports.LogLevel = logger_1.LogLevel;
var object_utils_1 = require("./lib/object-utils");
exports.assign = object_utils_1.assign;
exports.fillNulls = object_utils_1.fillNulls;
var guid_1 = require("./lib/guid");
exports.ensureGuid = guid_1.ensureGuid;
exports.initializeGuid = guid_1.initializeGuid;
exports.HasGuid = guid_1.HasGuid;
var collections_1 = require("./lib/collections");
exports.Stack = collections_1.Stack;
exports.Dict = collections_1.Dict;
exports.Set = collections_1.Set;
exports.DictSet = collections_1.DictSet;
exports.dict = collections_1.dict;
var list_utils_1 = require("./lib/list-utils");
exports.EMPTY_SLICE = list_utils_1.EMPTY_SLICE;
exports.LinkedList = list_utils_1.LinkedList;
exports.LinkedListNode = list_utils_1.LinkedListNode;
exports.ListNode = list_utils_1.ListNode;
exports.CloneableListNode = list_utils_1.CloneableListNode;
exports.ListSlice = list_utils_1.ListSlice;
exports.Slice = list_utils_1.Slice;
var array_utils_1 = require("./lib/array-utils");
exports.A = array_utils_1.default;
var weakmap_1 = require("./lib/weakmap");
exports.HAS_NATIVE_WEAKMAP = weakmap_1.HAS_NATIVE_WEAKMAP;
