// @glimmer/runtime/lib/syntax/specialize.js
"use strict";
var WireFormat = require("@glimmer/wire-format");
var util_1 = require("@glimmer/util");
var Specialize = (function () {
    function Specialize() {
        this.names = util_1.dict();
        this.funcs = [];
    }
    Specialize.prototype.add = function (name, func) {
        this.funcs.push(func);
        this.names[name] = this.funcs.length - 1;
    };
    Specialize.prototype.specialize = function (sexp, table) {
        var name = sexp[0];
        var index = this.names[name];
        if (index === undefined)
            return sexp;
        var func = this.funcs[index];
        util_1.assert(!!func, "expected a specialization for " + sexp[0]);
        return func(sexp, table);
    };
    return Specialize;
}());
exports.Specialize = Specialize;
exports.SPECIALIZE = new Specialize();
var E = WireFormat.Expressions;
exports.SPECIALIZE.add('append', function (sexp, _symbolTable) {
    var path = sexp[1];
    if (Array.isArray(path) && (E.isUnknown(path) || E.isGet(path))) {
        if (path[1].length !== 1) {
            return ['unoptimized-append', sexp[1], sexp[2]];
        }
    }
    return ['optimized-append', sexp[1], sexp[2]];
});
exports.SPECIALIZE.add('dynamic-attr', function (sexp, _symbolTable) {
    return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], false];
});
exports.SPECIALIZE.add('trusting-attr', function (sexp, _symbolTable) {
    return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], true];
});
exports.SPECIALIZE.add('partial', function (sexp, _table) {
    var expression = sexp[1];
    if (typeof expression === 'string') {
        return ['static-partial', expression];
    }
    else {
        return ['dynamic-partial', expression];
    }
});
