// glimmer-engine/packages/@glimmer/object/dist/commonjs/lib/descriptors.js
"use strict";
const mixin_1 = require("./mixin");
const computed_1 = require("./computed");
class AliasMethodDescriptor extends mixin_1.Descriptor {
    constructor(name) {
        super();
        this.name = name;
    }
    define(target, key, home) {
        let name = this.name;
        Object.defineProperty(target, key, {
            enumerable: true,
            configurable: true,
            get() {
                return this[name];
            }
        });
    }
}
class AliasMethodBlueprint extends mixin_1.Blueprint {
    constructor(name) {
        super();
        this.name = name;
    }
    descriptor(target, key, meta) {
        return new AliasMethodDescriptor(this.name);
    }
}
function aliasMethod(name) {
    return new AliasMethodBlueprint(name);
}
exports.aliasMethod = aliasMethod;
class AliasBlueprint extends computed_1.ComputedBlueprint {
    constructor(name) {
        let parent = name.slice(0, -1);
        let last = name[name.length - 1];
        let get = function () {
            return name.reduce((obj, n) => obj[n], this);
        };
        let set = function (value) {
            let p = parent.reduce((obj, n) => obj[n], this);
            p[last] = value;
        };
        super({ get, set }, [name]);
        this.name = name;
    }
    descriptor(target, key, meta) {
        if (this.name[0] === key)
            throw new Error(`Setting alias '${key}' on self`);
        return super.descriptor(target, key, meta);
    }
}
function alias(name) {
    return new AliasBlueprint(name.split('.'));
}
exports.alias = alias;
