// glimmer-engine/packages/glimmer-runtime/dist/commonjs/lib/syntax/specialize.js
"use strict";
const WireFormat = require("glimmer-wire-format");
const glimmer_util_1 = require("glimmer-util");
class Specialize {
    constructor() {
        this.names = glimmer_util_1.dict();
        this.funcs = [];
    }
    add(name, func) {
        this.funcs.push(func);
        this.names[name] = this.funcs.length - 1;
    }
    specialize(sexp, table) {
        let name = sexp[0];
        let index = this.names[name];
        if (index === undefined)
            return sexp;
        let func = this.funcs[index];
        glimmer_util_1.assert(!!func, `expected a specialization for ${sexp[0]}`);
        return func(sexp, table);
    }
}
exports.Specialize = Specialize;
exports.SPECIALIZE = new Specialize();
var E = WireFormat.Expressions;
exports.SPECIALIZE.add('append', (sexp, _symbolTable) => {
    let path = sexp[1];
    if (Array.isArray(path) && (E.isUnknown(path) || E.isGet(path))) {
        if (path[1].length !== 1) {
            return ['unoptimized-append', sexp[1], sexp[2]];
        }
    }
    return ['optimized-append', sexp[1], sexp[2]];
});
exports.SPECIALIZE.add('dynamic-attr', (sexp, _symbolTable) => {
    return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], false];
});
exports.SPECIALIZE.add('trusting-attr', (sexp, _symbolTable) => {
    return ['any-dynamic-attr', sexp[1], sexp[2], sexp[3], true];
});
exports.SPECIALIZE.add('partial', (sexp, _table) => {
    let expression = sexp[1];
    if (typeof expression === 'string') {
        return ['static-partial', expression];
    }
    else {
        return ['dynamic-partial', expression];
    }
});
