// glimmer-demos/lib/conways.js
"use strict";
var test_helpers_1 = require("@glimmer/test-helpers");
var object_reference_1 = require("@glimmer/object-reference");
// // Bare version
// const app = `{{#each world.cells key="key" as |cell|}}<organism-cell class="{{if cell.isAlive "alive" ""}}" style="top: {{cell.y}}0px; left: {{cell.x}}0px"/>{{/each}}`;
// Component version
var app = "<div class=\"world\">{{#each world.cells key=\"key\" as |cell|}}<organism-cell-component cell={{cell}} />{{/each}}</div>";
var env = new test_helpers_1.TestEnvironment();
env.registerEmberishGlimmerComponent('organism-cell-component', null, "<organism-cell class=\"{{if @cell.isAlive \"alive\" \"\"}}\" style=\"top: {{@cell.y}}px; left: {{@cell.x}}px\" />");
function getWorld() {
    /* tslint:disable: no-require-imports */
    return require("worlds/one").default;
    /* tslint:enable: no-require-imports */
}
var res;
var self;
function startGlimmer() {
    var world = getWorld();
    env.begin();
    self = new object_reference_1.UpdatableReference({ world: world });
    res = env.compile(app).render(self, document.body, new test_helpers_1.TestDynamicScope());
    env.commit();
    requestAnimationFrame(rerenderGlimmer);
}
exports.startGlimmer = startGlimmer;
function rerenderGlimmer() {
    var world = getWorld();
    world.advance();
    self.update({ world: world });
    res.rerender();
    requestAnimationFrame(rerenderGlimmer);
}
function startDOM() {
    var world = getWorld();
    var state = {
        isAlive: [],
        elements: []
    };
    var body = document.body;
    var div = document.createElement("div");
    div.classList.add("world");
    world.forEach(function (cell, i) {
        state.isAlive[i] = cell.isAlive;
        var el = state.elements[i] = document.createElement("organism-cell");
        el.style.top = cell.y + "px";
        el.style.left = cell.x + "px";
        if (cell.isAlive) {
            el.classList.add("alive");
        }
        div.appendChild(el);
    });
    body.appendChild(div);
    function updateCell(cell, i) {
        var oldAlive = state.isAlive[i];
        var newAlive = cell.isAlive;
        var el = state.elements[i];
        if (oldAlive !== newAlive) {
            state.isAlive[i] = newAlive;
            newAlive ? el.classList.add("alive") : el.classList.remove("alive");
        }
    }
    function rerenderDOM() {
        world.advance();
        world.forEach(updateCell);
        requestAnimationFrame(rerenderDOM);
    }
    requestAnimationFrame(rerenderDOM);
}
exports.startDOM = startDOM;
window['startDOM'] = startDOM;
