// @glimmer/runtime/lib/helpers/get-dynamic-var.js
"use strict";
var reference_1 = require("@glimmer/reference");
var DynamicVarReference = (function () {
    function DynamicVarReference(scope, nameRef) {
        this.scope = scope;
        this.nameRef = nameRef;
        var varTag = this.varTag = new reference_1.UpdatableTag(reference_1.CONSTANT_TAG);
        this.tag = reference_1.combine([nameRef.tag, varTag]);
    }
    DynamicVarReference.prototype.value = function () {
        return this.getVar().value();
    };
    DynamicVarReference.prototype.get = function (key) {
        return this.getVar().get(key);
    };
    DynamicVarReference.prototype.getVar = function () {
        var name = String(this.nameRef.value());
        var ref = this.scope.get(name);
        this.varTag.update(ref.tag);
        return ref;
    };
    return DynamicVarReference;
}());
function getDynamicVar(vm, args, _symbolTable) {
    var scope = vm.dynamicScope();
    var nameRef = args.positional.at(0);
    return new DynamicVarReference(scope, nameRef);
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = getDynamicVar;
