// glimmer-engine/dist/es6/@glimmer/runtime/lib/compiled/expressions/args.js
import { COMPILED_EMPTY_POSITIONAL_ARGS, EVALUATED_EMPTY_POSITIONAL_ARGS, CompiledPositionalArgs, EvaluatedPositionalArgs } from './positional-args';
import { COMPILED_EMPTY_NAMED_ARGS, EVALUATED_EMPTY_NAMED_ARGS, CompiledNamedArgs, EvaluatedNamedArgs } from './named-args';
import { combineTagged } from '@glimmer/reference';
export { COMPILED_EMPTY_POSITIONAL_ARGS, COMPILED_EMPTY_NAMED_ARGS };
export const EMPTY_BLOCKS = {
    default: null,
    inverse: null
};
export class CompiledArgs {
    constructor(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.type = "compiled-args";
    }
    static create(positional, named, blocks) {
        if (positional === COMPILED_EMPTY_POSITIONAL_ARGS && named === COMPILED_EMPTY_NAMED_ARGS && blocks === EMPTY_BLOCKS) {
            return this.empty();
        }
        else {
            return new this(positional, named, blocks);
        }
    }
    static empty() {
        return COMPILED_EMPTY_ARGS;
    }
    evaluate(vm) {
        let { positional, named, blocks } = this;
        return EvaluatedArgs.create(positional.evaluate(vm), named.evaluate(vm), blocks);
    }
}
export const COMPILED_EMPTY_ARGS = new (class extends CompiledArgs {
    constructor() {
        super(COMPILED_EMPTY_POSITIONAL_ARGS, COMPILED_EMPTY_NAMED_ARGS, EMPTY_BLOCKS);
    }
    evaluate(_vm) {
        return EMPTY_EVALUATED_ARGS;
    }
});
export class EvaluatedArgs {
    constructor(positional, named, blocks) {
        this.positional = positional;
        this.named = named;
        this.blocks = blocks;
        this.tag = combineTagged([positional, named]);
    }
    static empty() {
        return EMPTY_EVALUATED_ARGS;
    }
    static create(positional, named, blocks) {
        return new this(positional, named, blocks);
    }
    static positional(values, blocks = EMPTY_BLOCKS) {
        return new this(EvaluatedPositionalArgs.create(values), EVALUATED_EMPTY_NAMED_ARGS, blocks);
    }
    static named(map, blocks = EMPTY_BLOCKS) {
        return new this(EVALUATED_EMPTY_POSITIONAL_ARGS, EvaluatedNamedArgs.create(map), blocks);
    }
}
const EMPTY_EVALUATED_ARGS = new EvaluatedArgs(EVALUATED_EMPTY_POSITIONAL_ARGS, EVALUATED_EMPTY_NAMED_ARGS, EMPTY_BLOCKS);
export { CompiledPositionalArgs, EvaluatedPositionalArgs, CompiledNamedArgs, EvaluatedNamedArgs };
