// glimmer-engine/dist/es6/@glimmer/runtime/lib/compiled/expressions/has-block.js
import { CompiledExpression } from '../expressions';
import { PrimitiveReference } from '../../references';
export default class CompiledHasBlock extends CompiledExpression {
    constructor(inner) {
        super();
        this.inner = inner;
        this.type = "has-block";
    }
    evaluate(vm) {
        let block = this.inner.evaluate(vm);
        return PrimitiveReference.create(!!block);
    }
    toJSON() {
        return `has-block(${this.inner.toJSON()})`;
    }
}
export class CompiledHasBlockParams extends CompiledExpression {
    constructor(inner) {
        super();
        this.inner = inner;
        this.type = "has-block-params";
    }
    evaluate(vm) {
        let block = this.inner.evaluate(vm);
        let hasLocals = block && block.symbolTable.getSymbols().locals;
        return PrimitiveReference.create(!!hasLocals);
    }
    toJSON() {
        return `has-block-params(${this.inner.toJSON()})`;
    }
}
export class CompiledGetBlockBySymbol {
    constructor(symbol, debug) {
        this.symbol = symbol;
        this.debug = debug;
    }
    evaluate(vm) {
        return vm.scope().getBlock(this.symbol);
    }
    toJSON() {
        return `get-block($${this.symbol}(${this.debug}))`;
    }
}
export class CompiledInPartialGetBlock {
    constructor(symbol, name) {
        this.symbol = symbol;
        this.name = name;
    }
    evaluate(vm) {
        let { symbol, name } = this;
        let args = vm.scope().getPartialArgs(symbol);
        return args.blocks[name];
    }
    toJSON() {
        return `get-block($${this.symbol}($ARGS).${this.name}))`;
    }
}
