// glimmer-engine/dist/es6/@glimmer/runtime/lib/compiled/expressions/lookups.js
import { CompiledExpression } from '../expressions';
import { referenceFromParts } from '@glimmer/reference';
export default class CompiledLookup extends CompiledExpression {
    constructor(base, path) {
        super();
        this.base = base;
        this.path = path;
        this.type = "lookup";
    }
    static create(base, path) {
        if (path.length === 0) {
            return base;
        }
        else {
            return new this(base, path);
        }
    }
    evaluate(vm) {
        let { base, path } = this;
        return referenceFromParts(base.evaluate(vm), path);
    }
    toJSON() {
        return `${this.base.toJSON()}.${this.path.join('.')}`;
    }
}
export class CompiledSelf extends CompiledExpression {
    evaluate(vm) {
        return vm.getSelf();
    }
    toJSON() {
        return 'self';
    }
}
export class CompiledSymbol extends CompiledExpression {
    constructor(symbol, debug) {
        super();
        this.symbol = symbol;
        this.debug = debug;
    }
    evaluate(vm) {
        return vm.referenceForSymbol(this.symbol);
    }
    toJSON() {
        return `$${this.symbol}(${this.debug})`;
    }
}
export class CompiledInPartialName extends CompiledExpression {
    constructor(symbol, name) {
        super();
        this.symbol = symbol;
        this.name = name;
    }
    evaluate(vm) {
        let { symbol, name } = this;
        let args = vm.scope().getPartialArgs(symbol);
        return args.named.get(name);
    }
    toJSON() {
        return `$${this.symbol}($ARGS).${this.name}`;
    }
}
