// glimmer-engine/dist/es6/@glimmer/runtime/lib/compiled/expressions/positional-args.js
import { UNDEFINED_REFERENCE } from '../../references';
import { EMPTY_ARRAY } from '../../utils';
import { combineTagged } from '@glimmer/reference';
export class CompiledPositionalArgs {
    constructor(values) {
        this.values = values;
        this.length = values.length;
    }
    static create(values) {
        if (values.length) {
            return new this(values);
        }
        else {
            return COMPILED_EMPTY_POSITIONAL_ARGS;
        }
    }
    static empty() {
        return COMPILED_EMPTY_POSITIONAL_ARGS;
    }
    evaluate(vm) {
        let { values, length } = this;
        let references = new Array(length);
        for (let i = 0; i < length; i++) {
            references[i] = values[i].evaluate(vm);
        }
        return EvaluatedPositionalArgs.create(references);
    }
    toJSON() {
        return `[${this.values.map(value => value.toJSON()).join(", ")}]`;
    }
}
export const COMPILED_EMPTY_POSITIONAL_ARGS = new (class extends CompiledPositionalArgs {
    constructor() {
        super(EMPTY_ARRAY);
    }
    evaluate(_vm) {
        return EVALUATED_EMPTY_POSITIONAL_ARGS;
    }
    toJSON() {
        return `<EMPTY>`;
    }
});
export class EvaluatedPositionalArgs {
    constructor(values) {
        this.values = values;
        this.tag = combineTagged(values);
        this.length = values.length;
    }
    static create(values) {
        return new this(values);
    }
    static empty() {
        return EVALUATED_EMPTY_POSITIONAL_ARGS;
    }
    at(index) {
        let { values, length } = this;
        return (index < length) ? values[index] : UNDEFINED_REFERENCE;
    }
    value() {
        let { values, length } = this;
        let ret = new Array(length);
        for (let i = 0; i < length; i++) {
            ret[i] = values[i].value();
        }
        return ret;
    }
}
export const EVALUATED_EMPTY_POSITIONAL_ARGS = new (class extends EvaluatedPositionalArgs {
    constructor() {
        super(EMPTY_ARRAY);
    }
    at() {
        return UNDEFINED_REFERENCE;
    }
    value() {
        return this.values;
    }
});
