// @glimmer/runtime/lib/bounds.js
"use strict";
var Cursor = (function () {
    function Cursor(element, nextSibling) {
        this.element = element;
        this.nextSibling = nextSibling;
    }
    return Cursor;
}());
exports.Cursor = Cursor;
var RealDOMBounds = (function () {
    function RealDOMBounds(bounds) {
        this.bounds = bounds;
    }
    RealDOMBounds.prototype.parentElement = function () { return this.bounds.parentElement(); };
    RealDOMBounds.prototype.firstNode = function () { return this.bounds.firstNode(); };
    RealDOMBounds.prototype.lastNode = function () { return this.bounds.lastNode(); };
    return RealDOMBounds;
}());
exports.RealDOMBounds = RealDOMBounds;
var ConcreteBounds = (function () {
    function ConcreteBounds(parentNode, first, last) {
        this.parentNode = parentNode;
        this.first = first;
        this.last = last;
    }
    ConcreteBounds.prototype.parentElement = function () { return this.parentNode; };
    ConcreteBounds.prototype.firstNode = function () { return this.first; };
    ConcreteBounds.prototype.lastNode = function () { return this.last; };
    return ConcreteBounds;
}());
exports.ConcreteBounds = ConcreteBounds;
var SingleNodeBounds = (function () {
    function SingleNodeBounds(parentNode, node) {
        this.parentNode = parentNode;
        this.node = node;
    }
    SingleNodeBounds.prototype.parentElement = function () { return this.parentNode; };
    SingleNodeBounds.prototype.firstNode = function () { return this.node; };
    SingleNodeBounds.prototype.lastNode = function () { return this.node; };
    return SingleNodeBounds;
}());
exports.SingleNodeBounds = SingleNodeBounds;
function bounds(parent, first, last) {
    return new ConcreteBounds(parent, first, last);
}
exports.bounds = bounds;
function single(parent, node) {
    return new SingleNodeBounds(parent, node);
}
exports.single = single;
function move(bounds, reference) {
    var parent = bounds.parentElement();
    var first = bounds.firstNode();
    var last = bounds.lastNode();
    var node = first;
    while (node) {
        var next = node.nextSibling;
        parent.insertBefore(node, reference);
        if (node === last)
            return next;
        node = next;
    }
    return null;
}
exports.move = move;
function clear(bounds) {
    var parent = bounds.parentElement();
    var first = bounds.firstNode();
    var last = bounds.lastNode();
    var node = first;
    while (node) {
        var next = node.nextSibling;
        parent.removeChild(node);
        if (node === last)
            return next;
        node = next;
    }
    return null;
}
exports.clear = clear;
