// @glimmer/runtime/lib/references.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var reference_1 = require("@glimmer/reference");
var PrimitiveReference = (function (_super) {
    __extends(PrimitiveReference, _super);
    function PrimitiveReference(value) {
        return _super.call(this, value) || this;
    }
    PrimitiveReference.create = function (value) {
        if (value === undefined) {
            return exports.UNDEFINED_REFERENCE;
        }
        else if (value === null) {
            return exports.NULL_REFERENCE;
        }
        else if (value === true) {
            return TRUE_REFERENCE;
        }
        else if (value === false) {
            return FALSE_REFERENCE;
        }
        else if (typeof value === 'number') {
            return new ValueReference(value);
        }
        else {
            return new StringReference(value);
        }
    };
    PrimitiveReference.prototype.get = function (_key) {
        return exports.UNDEFINED_REFERENCE;
    };
    return PrimitiveReference;
}(reference_1.ConstReference));
exports.PrimitiveReference = PrimitiveReference;
var StringReference = (function (_super) {
    __extends(StringReference, _super);
    function StringReference() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.lengthReference = null;
        return _this;
    }
    StringReference.prototype.get = function (key) {
        if (key === 'length') {
            var lengthReference = this.lengthReference;
            if (lengthReference === null) {
                lengthReference = this.lengthReference = new ValueReference(this.inner.length);
            }
            return lengthReference;
        }
        else {
            return _super.prototype.get.call(this, key);
        }
    };
    return StringReference;
}(PrimitiveReference));
var ValueReference = (function (_super) {
    __extends(ValueReference, _super);
    function ValueReference(value) {
        return _super.call(this, value) || this;
    }
    return ValueReference;
}(PrimitiveReference));
exports.UNDEFINED_REFERENCE = new ValueReference(undefined);
exports.NULL_REFERENCE = new ValueReference(null);
var TRUE_REFERENCE = new ValueReference(true);
var FALSE_REFERENCE = new ValueReference(false);
var ConditionalReference = (function () {
    function ConditionalReference(inner) {
        this.inner = inner;
        this.tag = inner.tag;
    }
    ConditionalReference.prototype.value = function () {
        return this.toBool(this.inner.value());
    };
    ConditionalReference.prototype.toBool = function (value) {
        return !!value;
    };
    return ConditionalReference;
}());
exports.ConditionalReference = ConditionalReference;
