// glimmer-engine/dist/named-amd/@glimmer/object-reference/lib/references/descriptors.js
enifed('@glimmer/object-reference/lib/references/descriptors', ['exports', '@glimmer/object-reference/lib/meta', '@glimmer/reference'], function (exports, _glimmerObjectReferenceLibMeta, _glimmerReference) {
    'use strict';

    exports.ComputedReferenceBlueprint = ComputedReferenceBlueprint;

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var PropertyReference = (function () {
        function PropertyReference(object, property, outer) {
            _classCallCheck(this, PropertyReference);

            this.tag = _glimmerReference.VOLATILE_TAG;
            this.object = object;
            this.property = property;
        }

        PropertyReference.prototype.value = function value() {
            return this.object[this.property];
        };

        PropertyReference.prototype.label = function label() {
            return '[reference Property]';
        };

        return PropertyReference;
    })();

    exports.PropertyReference = PropertyReference;

    function ComputedReferenceBlueprint(property, dependencies) {
        return (function () {
            function ComputedReference(object, property, outer) {
                _classCallCheck(this, ComputedReference);

                this.installed = false;
                this.tag = _glimmerReference.VOLATILE_TAG;
                this.object = object;
                this.property = property;
                this.dependencies = dependencies;
                this.outer = outer;
            }

            ComputedReference.prototype.value = function value() {
                var _this = this;

                if (!this.installed) {
                    (function () {
                        var root = _glimmerObjectReferenceLibMeta.default.for(_this.object).root();
                        _this.dependencies.forEach(function (dep) {
                            var ref = root.referenceFromParts(dep);
                            ref.value();
                        });
                        _this.installed = true;
                    })();
                }
                return this.object[this.property];
            };

            ComputedReference.prototype.label = function label() {
                return '[reference Computed]';
            };

            return ComputedReference;
        })();
    }
});
