// glimmer-engine/dist/named-amd/@glimmer/object-reference/lib/references/path.js
enifed('@glimmer/object-reference/lib/references/path', ['exports', '@glimmer/object-reference/lib/utils', '@glimmer/util', '@glimmer/object-reference/lib/meta', '@glimmer/object-reference/lib/references/descriptors', '@glimmer/reference'], function (exports, _glimmerObjectReferenceLibUtils, _glimmerUtil, _glimmerObjectReferenceLibMeta, _glimmerObjectReferenceLibReferencesDescriptors, _glimmerReference) {
    'use strict';

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var UnchainFromPath = (function () {
        function UnchainFromPath(set, child) {
            _classCallCheck(this, UnchainFromPath);

            this.set = set;
            this.child = child;
        }

        UnchainFromPath.prototype.destroy = function destroy() {
            this.set.delete(this.child);
        };

        return UnchainFromPath;
    })();

    var PathReference = (function () {
        function PathReference(parent, property) {
            _classCallCheck(this, PathReference);

            this.cache = _glimmerObjectReferenceLibUtils.EMPTY_CACHE;
            this.inner = null;
            this.chains = null;
            this.lastParentValue = _glimmerObjectReferenceLibUtils.EMPTY_CACHE;
            this._guid = null;
            this.tag = _glimmerReference.VOLATILE_TAG;
            this.parent = parent;
            this.property = property;
        }

        PathReference.prototype.value = function value() {
            var lastParentValue = this.lastParentValue;
            var property = this.property;
            var inner = this.inner;

            var parentValue = this._parentValue();
            if (parentValue === null || parentValue === undefined) {
                return this.cache = undefined;
            }
            if (lastParentValue === parentValue) {
                inner = this.inner;
            } else {
                var ReferenceType = typeof parentValue === 'object' ? _glimmerObjectReferenceLibMeta.default.for(parentValue).referenceTypeFor(property) : _glimmerObjectReferenceLibReferencesDescriptors.PropertyReference;
                inner = this.inner = new ReferenceType(parentValue, property, this);
            }
            // if (typeof parentValue === 'object') {
            //   Meta.for(parentValue).addReference(property, this);
            // }
            return this.cache = inner.value();
        };

        PathReference.prototype.get = function get(prop) {
            var chains = this._getChains();
            if (prop in chains) return chains[prop];
            return chains[prop] = new PathReference(this, prop);
        };

        PathReference.prototype.label = function label() {
            return '[reference Direct]';
        };

        PathReference.prototype._getChains = function _getChains() {
            if (this.chains) return this.chains;
            return this.chains = _glimmerUtil.dict();
        };

        PathReference.prototype._parentValue = function _parentValue() {
            var parent = this.parent.value();
            this.lastParentValue = parent;
            return parent;
        };

        return PathReference;
    })();

    exports.default = PathReference;
    exports.PathReference = PathReference;
});
