// glimmer-engine/dist/named-amd/@glimmer/object-reference/lib/references/root.js
enifed('@glimmer/object-reference/lib/references/root', ['exports', '@glimmer/util', '@glimmer/object-reference/lib/references/path', '@glimmer/reference'], function (exports, _glimmerUtil, _glimmerObjectReferenceLibReferencesPath, _glimmerReference) {
    'use strict';

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

    var RootReference = (function () {
        function RootReference(object) {
            _classCallCheck(this, RootReference);

            this.chains = _glimmerUtil.dict();
            this.tag = _glimmerReference.VOLATILE_TAG;
            this.object = object;
        }

        RootReference.prototype.value = function value() {
            return this.object;
        };

        RootReference.prototype.update = function update(object) {
            this.object = object;
            // this.notify();
        };

        RootReference.prototype.get = function get(prop) {
            var chains = this.chains;
            if (prop in chains) return chains[prop];
            return chains[prop] = new _glimmerObjectReferenceLibReferencesPath.PathReference(this, prop);
        };

        RootReference.prototype.chainFor = function chainFor(prop) {
            var chains = this.chains;
            if (prop in chains) return chains[prop];
            return null;
        };

        RootReference.prototype.path = function path(string) {
            return string.split('.').reduce(function (ref, part) {
                return ref.get(part);
            }, this);
        };

        RootReference.prototype.referenceFromParts = function referenceFromParts(parts) {
            return parts.reduce(function (ref, part) {
                return ref.get(part);
            }, this);
        };

        RootReference.prototype.label = function label() {
            return '[reference Root]';
        };

        return RootReference;
    })();

    exports.default = RootReference;
});
