// @glimmer/runtime/lib/vm/append.js
"use strict";
var environment_1 = require("../environment");
var util_1 = require("@glimmer/util");
var reference_1 = require("@glimmer/reference");
var vm_1 = require("../compiled/opcodes/vm");
var update_1 = require("./update");
var render_result_1 = require("./render-result");
var frame_1 = require("./frame");
var opcodes_1 = require("../opcodes");
var VM = (function () {
    function VM(env, scope, dynamicScope, elementStack) {
        this.env = env;
        this.elementStack = elementStack;
        this.dynamicScopeStack = new util_1.Stack();
        this.scopeStack = new util_1.Stack();
        this.updatingOpcodeStack = new util_1.Stack();
        this.cacheGroups = new util_1.Stack();
        this.listBlockStack = new util_1.Stack();
        this.frame = new frame_1.FrameStack();
        this.env = env;
        this.constants = env.constants;
        this.elementStack = elementStack;
        this.scopeStack.push(scope);
        this.dynamicScopeStack.push(dynamicScope);
    }
    VM.initial = function (env, self, dynamicScope, elementStack, size) {
        var scope = environment_1.Scope.root(self, size);
        return new VM(env, scope, dynamicScope, elementStack);
    };
    VM.prototype.capture = function () {
        return {
            env: this.env,
            scope: this.scope(),
            dynamicScope: this.dynamicScope(),
            frame: this.frame.capture()
        };
    };
    VM.prototype.goto = function (ip) {
        // assert(this.frame.getOps().contains(op), `Illegal jump to ${op.label}`);
        this.frame.goto(ip);
    };
    VM.prototype.beginCacheGroup = function () {
        this.cacheGroups.push(this.updating().tail());
    };
    VM.prototype.commitCacheGroup = function () {
        //        JumpIfNotModified(END)
        //        (head)
        //        (....)
        //        (tail)
        //        DidModify
        // END:   Noop
        var END = new vm_1.LabelOpcode("END");
        var opcodes = this.updating();
        var marker = this.cacheGroups.pop();
        var head = marker ? opcodes.nextNode(marker) : opcodes.head();
        var tail = opcodes.tail();
        var tag = reference_1.combineSlice(new util_1.ListSlice(head, tail));
        var guard = new vm_1.JumpIfNotModifiedOpcode(tag, END);
        opcodes.insertBefore(guard, head);
        opcodes.append(new vm_1.DidModifyOpcode(guard));
        opcodes.append(END);
    };
    VM.prototype.enter = function (sliceId) {
        var updating = new util_1.LinkedList();
        var tracker = this.stack().pushUpdatableBlock();
        var state = this.capture();
        var slice = this.constants.getSlice(sliceId);
        var tryOpcode = new update_1.TryOpcode(slice, state, tracker, updating);
        this.didEnter(tryOpcode, updating);
    };
    VM.prototype.enterWithKey = function (key, ops) {
        var updating = new util_1.LinkedList();
        var tracker = this.stack().pushUpdatableBlock();
        var state = this.capture();
        var tryOpcode = new update_1.TryOpcode(ops, state, tracker, updating);
        this.listBlock().map[key] = tryOpcode;
        this.didEnter(tryOpcode, updating);
    };
    VM.prototype.enterList = function (ops) {
        var updating = new util_1.LinkedList();
        var tracker = this.stack().pushBlockList(updating);
        var state = this.capture();
        var artifacts = this.frame.getIterator().artifacts;
        var opcode = new update_1.ListBlockOpcode(ops, state, tracker, updating, artifacts);
        this.listBlockStack.push(opcode);
        this.didEnter(opcode, updating);
    };
    VM.prototype.didEnter = function (opcode, updating) {
        this.updateWith(opcode);
        this.updatingOpcodeStack.push(updating);
    };
    VM.prototype.exit = function () {
        this.stack().popBlock();
        this.updatingOpcodeStack.pop();
        var parent = this.updating().tail();
        parent.didInitializeChildren();
    };
    VM.prototype.exitList = function () {
        this.exit();
        this.listBlockStack.pop();
    };
    VM.prototype.updateWith = function (opcode) {
        this.updating().append(opcode);
    };
    VM.prototype.listBlock = function () {
        return util_1.expect(this.listBlockStack.current, 'expected a list block');
    };
    VM.prototype.updating = function () {
        return util_1.expect(this.updatingOpcodeStack.current, 'expected updating opcode on the updating opcode stack');
    };
    VM.prototype.stack = function () {
        return this.elementStack;
    };
    VM.prototype.scope = function () {
        return util_1.expect(this.scopeStack.current, 'expected scope on the scope stack');
    };
    VM.prototype.dynamicScope = function () {
        return util_1.expect(this.dynamicScopeStack.current, 'expected dynamic scope on the dynamic scope stack');
    };
    VM.prototype.pushFrame = function (block, args, callerScope) {
        this.frame.push(block.slice);
        if (args)
            this.frame.setArgs(args);
        if (args && args.blocks)
            this.frame.setBlocks(args.blocks);
        if (callerScope)
            this.frame.setCallerScope(callerScope);
    };
    VM.prototype.pushComponentFrame = function (layout, args, callerScope, component, manager, shadow) {
        this.frame.push(layout.slice, component, manager, shadow);
        if (args)
            this.frame.setArgs(args);
        if (args && args.blocks)
            this.frame.setBlocks(args.blocks);
        if (callerScope)
            this.frame.setCallerScope(callerScope);
    };
    VM.prototype.pushEvalFrame = function (slice) {
        this.frame.push(slice);
    };
    VM.prototype.pushChildScope = function () {
        this.scopeStack.push(this.scope().child());
    };
    VM.prototype.pushCallerScope = function () {
        this.scopeStack.push(util_1.expect(this.scope().getCallerScope(), 'pushCallerScope is called when a caller scope is present'));
    };
    VM.prototype.pushDynamicScope = function () {
        var child = this.dynamicScope().child();
        this.dynamicScopeStack.push(child);
        return child;
    };
    VM.prototype.pushRootScope = function (self, size) {
        var scope = environment_1.Scope.root(self, size);
        this.scopeStack.push(scope);
        return scope;
    };
    VM.prototype.popScope = function () {
        this.scopeStack.pop();
    };
    VM.prototype.popDynamicScope = function () {
        this.dynamicScopeStack.pop();
    };
    VM.prototype.newDestroyable = function (d) {
        this.stack().newDestroyable(d);
    };
    /// SCOPE HELPERS
    VM.prototype.getSelf = function () {
        return this.scope().getSelf();
    };
    VM.prototype.referenceForSymbol = function (symbol) {
        return this.scope().getSymbol(symbol);
    };
    VM.prototype.getArgs = function () {
        return this.frame.getArgs();
    };
    /// EXECUTION
    VM.prototype.resume = function (opcodes, frame) {
        return this.execute(opcodes, function (vm) { return vm.frame.restore(frame); });
    };
    VM.prototype.execute = function (opcodes, initialize) {
        util_1.LOGGER.debug("[VM] Begin program execution");
        var _a = this, elementStack = _a.elementStack, frame = _a.frame, updatingOpcodeStack = _a.updatingOpcodeStack, env = _a.env;
        elementStack.pushSimpleBlock();
        updatingOpcodeStack.push(new util_1.LinkedList());
        frame.push(opcodes);
        if (initialize)
            initialize(this);
        var opcode;
        while (frame.hasOpcodes()) {
            if (opcode = frame.nextStatement(this.env)) {
                util_1.LOGGER.trace(opcode);
                opcodes_1.APPEND_OPCODES.evaluate(this, opcode);
            }
        }
        util_1.LOGGER.debug("[VM] Completed program execution");
        return new render_result_1.default(env, util_1.expect(updatingOpcodeStack.pop(), 'there should be a final updating opcode stack'), elementStack.popBlock());
    };
    VM.prototype.evaluateOpcode = function (opcode) {
        opcodes_1.APPEND_OPCODES.evaluate(this, opcode);
    };
    // Make sure you have opcodes that push and pop a scope around this opcode
    // if you need to change the scope.
    VM.prototype.invokeBlock = function (block, args) {
        var compiled = block.compile(this.env);
        this.pushFrame(compiled, args);
    };
    VM.prototype.invokePartial = function (block) {
        var compiled = block.compile(this.env);
        this.pushFrame(compiled);
    };
    VM.prototype.invokeLayout = function (args, layout, callerScope, component, manager, shadow) {
        this.pushComponentFrame(layout, args, callerScope, component, manager, shadow);
    };
    VM.prototype.evaluateOperand = function (expr) {
        this.frame.setOperand(expr.evaluate(this));
    };
    VM.prototype.evaluateArgs = function (args) {
        var evaledArgs = this.frame.setArgs(args.evaluate(this));
        this.frame.setOperand(evaledArgs.positional.at(0));
    };
    VM.prototype.bindPositionalArgs = function (symbols) {
        var args = util_1.expect(this.frame.getArgs(), 'bindPositionalArgs assumes a previous setArgs');
        var positional = args.positional;
        var scope = this.scope();
        for (var i = 0; i < symbols.length; i++) {
            scope.bindSymbol(symbols[i], positional.at(i));
        }
    };
    VM.prototype.bindNamedArgs = function (names, symbols) {
        var args = util_1.expect(this.frame.getArgs(), 'bindNamedArgs assumes a previous setArgs');
        var scope = this.scope();
        var named = args.named;
        for (var i = 0; i < names.length; i++) {
            var name = this.constants.getString(names[i]);
            scope.bindSymbol(symbols[i], named.get(name));
        }
    };
    VM.prototype.bindBlocks = function (names, symbols) {
        var blocks = this.frame.getBlocks();
        var scope = this.scope();
        for (var i = 0; i < names.length; i++) {
            var name = this.constants.getString(names[i]);
            scope.bindBlock(symbols[i], (blocks && blocks[name]) || null);
        }
    };
    VM.prototype.bindPartialArgs = function (symbol) {
        var args = util_1.expect(this.frame.getArgs(), 'bindPartialArgs assumes a previous setArgs');
        var scope = this.scope();
        util_1.assert(args, "Cannot bind named args");
        scope.bindPartialArgs(symbol, args);
    };
    VM.prototype.bindCallerScope = function () {
        var callerScope = this.frame.getCallerScope();
        var scope = this.scope();
        util_1.assert(callerScope, "Cannot bind caller scope");
        scope.bindCallerScope(callerScope);
    };
    VM.prototype.bindDynamicScope = function (names) {
        var args = util_1.expect(this.frame.getArgs(), 'bindDynamicScope assumes a previous setArgs');
        var scope = this.dynamicScope();
        util_1.assert(args, "Cannot bind dynamic scope");
        for (var i = 0; i < names.length; i++) {
            var name = this.constants.getString(names[i]);
            scope.set(name, args.named.get(name));
        }
    };
    return VM;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = VM;
