// @glimmer/runtime/lib/vm/render-result.js
"use strict";
var bounds_1 = require("../bounds");
var update_1 = require("./update");
var RenderResult = (function () {
    function RenderResult(env, updating, bounds) {
        this.env = env;
        this.updating = updating;
        this.bounds = bounds;
    }
    RenderResult.prototype.rerender = function (_a) {
        var _b = (_a === void 0 ? { alwaysRevalidate: false } : _a).alwaysRevalidate, alwaysRevalidate = _b === void 0 ? false : _b;
        var _c = this, env = _c.env, updating = _c.updating;
        var vm = new update_1.default(env, { alwaysRevalidate: alwaysRevalidate });
        vm.execute(updating, this);
    };
    RenderResult.prototype.parentElement = function () {
        return this.bounds.parentElement();
    };
    RenderResult.prototype.firstNode = function () {
        return this.bounds.firstNode();
    };
    RenderResult.prototype.lastNode = function () {
        return this.bounds.lastNode();
    };
    RenderResult.prototype.opcodes = function () {
        return this.updating;
    };
    RenderResult.prototype.handleException = function () {
        throw "this should never happen";
    };
    RenderResult.prototype.destroy = function () {
        this.bounds.destroy();
        bounds_1.clear(this.bounds);
    };
    return RenderResult;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RenderResult;
