// glimmer-engine/dist/named-amd/@glimmer/runtime/lib/compiled/blocks.js
enifed("@glimmer/runtime/lib/compiled/blocks", ["exports"], function (exports) {
    "use strict";

    function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

    function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

    function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

    var CompiledBlock = function CompiledBlock(slice) {
        _classCallCheck(this, CompiledBlock);

        this.start = slice[0];
        this.end = slice[1];
        this.slice = slice;
    };

    exports.CompiledBlock = CompiledBlock;

    var CompiledProgram = (function (_CompiledBlock) {
        _inherits(CompiledProgram, _CompiledBlock);

        function CompiledProgram(slice, symbols) {
            _classCallCheck(this, CompiledProgram);

            _CompiledBlock.call(this, slice);
            this.symbols = symbols;
        }

        return CompiledProgram;
    })(CompiledBlock);

    exports.CompiledProgram = CompiledProgram;
});
