// @glimmer/object-reference/lib/references/descriptors.js
"use strict";
var meta_1 = require("../meta");
var reference_1 = require("@glimmer/reference");
var PropertyReference = (function () {
    function PropertyReference(object, property, outer) {
        this.tag = reference_1.VOLATILE_TAG;
        this.object = object;
        this.property = property;
    }
    PropertyReference.prototype.value = function () { return this.object[this.property]; };
    PropertyReference.prototype.label = function () {
        return '[reference Property]';
    };
    return PropertyReference;
}());
exports.PropertyReference = PropertyReference;
function ComputedReferenceBlueprint(property, dependencies) {
    return (function () {
        function ComputedReference(object, property, outer) {
            this.installed = false;
            this.tag = reference_1.VOLATILE_TAG;
            this.object = object;
            this.property = property;
            this.dependencies = dependencies;
            this.outer = outer;
        }
        ComputedReference.prototype.value = function () {
            if (!this.installed) {
                var root_1 = meta_1.default.for(this.object).root();
                this.dependencies.forEach(function (dep) {
                    var ref = root_1.referenceFromParts(dep);
                    ref.value();
                });
                this.installed = true;
            }
            return this.object[this.property];
        };
        ComputedReference.prototype.label = function () {
            return '[reference Computed]';
        };
        return ComputedReference;
    }());
}
exports.ComputedReferenceBlueprint = ComputedReferenceBlueprint;
