// @glimmer/object-reference/lib/references/path.js
"use strict";
var utils_1 = require("../utils");
var util_1 = require("@glimmer/util");
var meta_1 = require("../meta");
var descriptors_1 = require("./descriptors");
var reference_1 = require("@glimmer/reference");
var UnchainFromPath = (function () {
    function UnchainFromPath(set, child) {
        this.set = set;
        this.child = child;
    }
    UnchainFromPath.prototype.destroy = function () {
        this.set.delete(this.child);
    };
    return UnchainFromPath;
}());
var PathReference = (function () {
    function PathReference(parent, property) {
        this.cache = utils_1.EMPTY_CACHE;
        this.inner = null;
        this.chains = null;
        this.lastParentValue = utils_1.EMPTY_CACHE;
        this._guid = null;
        this.tag = reference_1.VOLATILE_TAG;
        this.parent = parent;
        this.property = property;
    }
    PathReference.prototype.value = function () {
        var _a = this, lastParentValue = _a.lastParentValue, property = _a.property, inner = _a.inner;
        var parentValue = this._parentValue();
        if (parentValue === null || parentValue === undefined) {
            return (this.cache = undefined);
        }
        if (lastParentValue === parentValue) {
            inner = this.inner;
        }
        else {
            var ReferenceType = typeof parentValue === 'object' ? meta_1.default.for(parentValue).referenceTypeFor(property) : descriptors_1.PropertyReference;
            inner = this.inner = new ReferenceType(parentValue, property, this);
        }
        // if (typeof parentValue === 'object') {
        //   Meta.for(parentValue).addReference(property, this);
        // }
        return (this.cache = inner.value());
    };
    PathReference.prototype.get = function (prop) {
        var chains = this._getChains();
        if (prop in chains)
            return chains[prop];
        return (chains[prop] = new PathReference(this, prop));
    };
    PathReference.prototype.label = function () {
        return '[reference Direct]';
    };
    PathReference.prototype._getChains = function () {
        if (this.chains)
            return this.chains;
        return (this.chains = util_1.dict());
    };
    PathReference.prototype._parentValue = function () {
        var parent = this.parent.value();
        this.lastParentValue = parent;
        return parent;
    };
    return PathReference;
}());
exports.PathReference = PathReference;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PathReference;
