// @glimmer/object-reference/lib/references/root.js
"use strict";
var util_1 = require("@glimmer/util");
var path_1 = require("./path");
var reference_1 = require("@glimmer/reference");
var RootReference = (function () {
    function RootReference(object) {
        this.chains = util_1.dict();
        this.tag = reference_1.VOLATILE_TAG;
        this.object = object;
    }
    RootReference.prototype.value = function () { return this.object; };
    RootReference.prototype.update = function (object) {
        this.object = object;
        // this.notify();
    };
    RootReference.prototype.get = function (prop) {
        var chains = this.chains;
        if (prop in chains)
            return chains[prop];
        return (chains[prop] = new path_1.PathReference(this, prop));
    };
    RootReference.prototype.chainFor = function (prop) {
        var chains = this.chains;
        if (prop in chains)
            return chains[prop];
        return null;
    };
    RootReference.prototype.path = function (string) {
        return string.split('.').reduce(function (ref, part) { return ref.get(part); }, this);
    };
    RootReference.prototype.referenceFromParts = function (parts) {
        return parts.reduce(function (ref, part) { return ref.get(part); }, this);
    };
    RootReference.prototype.label = function () {
        return '[reference Root]';
    };
    return RootReference;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = RootReference;
