// glimmer-engine/packages/glimmer-runtime/dist/commonjs/lib/compiled/expressions/has-block.js
"use strict";
const expressions_1 = require("../expressions");
const references_1 = require("../../references");
class CompiledHasBlock extends expressions_1.CompiledExpression {
    constructor(inner) {
        super();
        this.inner = inner;
        this.type = "has-block";
    }
    evaluate(vm) {
        let block = this.inner.evaluate(vm);
        return references_1.PrimitiveReference.create(!!block);
    }
    toJSON() {
        return `has-block(${this.inner.toJSON()})`;
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledHasBlock;
class CompiledHasBlockParams extends expressions_1.CompiledExpression {
    constructor(inner) {
        super();
        this.inner = inner;
        this.type = "has-block-params";
    }
    evaluate(vm) {
        let block = this.inner.evaluate(vm);
        let hasLocals = block && block.symbolTable.getSymbols().locals;
        return references_1.PrimitiveReference.create(!!hasLocals);
    }
    toJSON() {
        return `has-block-params(${this.inner.toJSON()})`;
    }
}
exports.CompiledHasBlockParams = CompiledHasBlockParams;
class CompiledGetBlockBySymbol {
    constructor(symbol, debug) {
        this.symbol = symbol;
        this.debug = debug;
    }
    evaluate(vm) {
        return vm.scope().getBlock(this.symbol);
    }
    toJSON() {
        return `get-block($${this.symbol}(${this.debug}))`;
    }
}
exports.CompiledGetBlockBySymbol = CompiledGetBlockBySymbol;
class CompiledInPartialGetBlock {
    constructor(symbol, name) {
        this.symbol = symbol;
        this.name = name;
    }
    evaluate(vm) {
        let { symbol, name } = this;
        let args = vm.scope().getPartialArgs(symbol);
        return args.blocks[name];
    }
    toJSON() {
        return `get-block($${this.symbol}($ARGS).${this.name}))`;
    }
}
exports.CompiledInPartialGetBlock = CompiledInPartialGetBlock;
