// glimmer-engine/packages/glimmer-runtime/dist/commonjs/lib/compiled/expressions/lookups.js
"use strict";
const expressions_1 = require("../expressions");
const glimmer_reference_1 = require("glimmer-reference");
class CompiledLookup extends expressions_1.CompiledExpression {
    constructor(base, path) {
        super();
        this.base = base;
        this.path = path;
        this.type = "lookup";
    }
    static create(base, path) {
        if (path.length === 0) {
            return base;
        }
        else {
            return new this(base, path);
        }
    }
    evaluate(vm) {
        let { base, path } = this;
        return glimmer_reference_1.referenceFromParts(base.evaluate(vm), path);
    }
    toJSON() {
        return `${this.base.toJSON()}.${this.path.join('.')}`;
    }
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CompiledLookup;
class CompiledSelf extends expressions_1.CompiledExpression {
    evaluate(vm) {
        return vm.getSelf();
    }
    toJSON() {
        return 'self';
    }
}
exports.CompiledSelf = CompiledSelf;
class CompiledSymbol extends expressions_1.CompiledExpression {
    constructor(symbol, debug) {
        super();
        this.symbol = symbol;
        this.debug = debug;
    }
    evaluate(vm) {
        return vm.referenceForSymbol(this.symbol);
    }
    toJSON() {
        return `$${this.symbol}(${this.debug})`;
    }
}
exports.CompiledSymbol = CompiledSymbol;
class CompiledInPartialName extends expressions_1.CompiledExpression {
    constructor(symbol, name) {
        super();
        this.symbol = symbol;
        this.name = name;
    }
    evaluate(vm) {
        let { symbol, name } = this;
        let args = vm.scope().getPartialArgs(symbol);
        return args.named.get(name);
    }
    toJSON() {
        return `$${this.symbol}($ARGS).${this.name}`;
    }
}
exports.CompiledInPartialName = CompiledInPartialName;
