// glimmer-engine/packages/glimmer-runtime/dist/commonjs/lib/compiled/expressions/positional-args.js
"use strict";
const references_1 = require("../../references");
const utils_1 = require("../../utils");
const glimmer_reference_1 = require("glimmer-reference");
class CompiledPositionalArgs {
    constructor(values) {
        this.values = values;
        this.length = values.length;
    }
    static create(values) {
        if (values.length) {
            return new this(values);
        }
        else {
            return exports.COMPILED_EMPTY_POSITIONAL_ARGS;
        }
    }
    static empty() {
        return exports.COMPILED_EMPTY_POSITIONAL_ARGS;
    }
    evaluate(vm) {
        let { values, length } = this;
        let references = new Array(length);
        for (let i = 0; i < length; i++) {
            references[i] = values[i].evaluate(vm);
        }
        return EvaluatedPositionalArgs.create(references);
    }
    toJSON() {
        return `[${this.values.map(value => value.toJSON()).join(", ")}]`;
    }
}
exports.CompiledPositionalArgs = CompiledPositionalArgs;
exports.COMPILED_EMPTY_POSITIONAL_ARGS = new (class extends CompiledPositionalArgs {
    constructor() {
        super(utils_1.EMPTY_ARRAY);
    }
    evaluate(_vm) {
        return exports.EVALUATED_EMPTY_POSITIONAL_ARGS;
    }
    toJSON() {
        return `<EMPTY>`;
    }
});
class EvaluatedPositionalArgs {
    constructor(values) {
        this.values = values;
        this.tag = glimmer_reference_1.combineTagged(values);
        this.length = values.length;
    }
    static create(values) {
        return new this(values);
    }
    static empty() {
        return exports.EVALUATED_EMPTY_POSITIONAL_ARGS;
    }
    at(index) {
        let { values, length } = this;
        return (index < length) ? values[index] : references_1.UNDEFINED_REFERENCE;
    }
    value() {
        let { values, length } = this;
        let ret = new Array(length);
        for (let i = 0; i < length; i++) {
            ret[i] = values[i].value();
        }
        return ret;
    }
}
exports.EvaluatedPositionalArgs = EvaluatedPositionalArgs;
exports.EVALUATED_EMPTY_POSITIONAL_ARGS = new (class extends EvaluatedPositionalArgs {
    constructor() {
        super(utils_1.EMPTY_ARRAY);
    }
    at() {
        return references_1.UNDEFINED_REFERENCE;
    }
    value() {
        return this.values;
    }
});
