// @glimmer/runtime/index.js
"use strict";
require("./lib/bootstrap");
var template_1 = require("./lib/template");
exports.templateFactory = template_1.default;
exports.TemplateFactory = template_1.TemplateFactory;
exports.Template = template_1.Template;
var references_1 = require("./lib/references");
exports.NULL_REFERENCE = references_1.NULL_REFERENCE;
exports.UNDEFINED_REFERENCE = references_1.UNDEFINED_REFERENCE;
exports.PrimitiveReference = references_1.PrimitiveReference;
exports.ConditionalReference = references_1.ConditionalReference;
var builder_1 = require("./lib/compiled/opcodes/builder");
exports.OpcodeBuilderDSL = builder_1.default;
var compiler_1 = require("./lib/compiler");
exports.CompilableLayout = compiler_1.CompilableLayout;
exports.compileLayout = compiler_1.compileLayout;
var opcode_builder_1 = require("./lib/opcode-builder");
exports.ComponentBuilder = opcode_builder_1.ComponentBuilder;
exports.StaticDefinition = opcode_builder_1.StaticDefinition;
exports.DynamicDefinition = opcode_builder_1.DynamicDefinition;
var blocks_1 = require("./lib/compiled/blocks");
exports.CompiledBlock = blocks_1.CompiledBlock;
exports.CompiledProgram = blocks_1.CompiledProgram;
var attribute_managers_1 = require("./lib/dom/attribute-managers");
exports.IAttributeManager = attribute_managers_1.AttributeManager;
exports.AttributeManager = attribute_managers_1.AttributeManager;
exports.PropertyManager = attribute_managers_1.PropertyManager;
exports.INPUT_VALUE_PROPERTY_MANAGER = attribute_managers_1.INPUT_VALUE_PROPERTY_MANAGER;
exports.defaultManagers = attribute_managers_1.defaultManagers;
exports.defaultAttributeManagers = attribute_managers_1.defaultAttributeManagers;
exports.defaultPropertyManagers = attribute_managers_1.defaultPropertyManagers;
exports.readDOMAttr = attribute_managers_1.readDOMAttr;
var content_1 = require("./lib/compiled/opcodes/content");
exports.normalizeTextValue = content_1.normalizeTextValue;
var expressions_1 = require("./lib/compiled/expressions");
exports.CompiledExpression = expressions_1.CompiledExpression;
var args_1 = require("./lib/compiled/expressions/args");
exports.CompiledArgs = args_1.CompiledArgs;
exports.CompiledNamedArgs = args_1.CompiledNamedArgs;
exports.CompiledPositionalArgs = args_1.CompiledPositionalArgs;
exports.EvaluatedArgs = args_1.EvaluatedArgs;
exports.EvaluatedNamedArgs = args_1.EvaluatedNamedArgs;
exports.EvaluatedPositionalArgs = args_1.EvaluatedPositionalArgs;
var function_1 = require("./lib/compiled/expressions/function");
exports.FunctionExpression = function_1.FunctionExpression;
var get_dynamic_var_1 = require("./lib/helpers/get-dynamic-var");
exports.getDynamicVar = get_dynamic_var_1.default;
var functions_1 = require("./lib/syntax/functions");
exports.BlockMacros = functions_1.Blocks;
exports.InlineMacros = functions_1.Inlines;
exports.NestedBlockSyntax = functions_1.NestedBlockSyntax;
exports.CompileBlockMacro = functions_1.CompileBlockMacro;
exports.compileArgs = functions_1.compileArgs;
exports.setDebuggerCallback = functions_1.setDebuggerCallback;
exports.resetDebuggerCallback = functions_1.resetDebuggerCallback;
exports.debugCallback = functions_1.debugCallback;
var scanner_1 = require("./lib/scanner");
exports.BaselineSyntax = scanner_1.BaselineSyntax;
exports.Layout = scanner_1.Layout;
var vm_1 = require("./lib/vm");
exports.VM = vm_1.PublicVM;
exports.UpdatingVM = vm_1.UpdatingVM;
exports.RenderResult = vm_1.RenderResult;
var upsert_1 = require("./lib/upsert");
exports.SafeString = upsert_1.SafeString;
exports.isSafeString = upsert_1.isSafeString;
var environment_1 = require("./lib/environment");
exports.Scope = environment_1.Scope;
exports.Environment = environment_1.default;
exports.Helper = environment_1.Helper;
exports.DynamicScope = environment_1.DynamicScope;
var partial_1 = require("./lib/partial");
exports.PartialDefinition = partial_1.PartialDefinition;
var interfaces_1 = require("./lib/component/interfaces");
exports.Component = interfaces_1.Component;
exports.ComponentClass = interfaces_1.ComponentClass;
exports.ComponentManager = interfaces_1.ComponentManager;
exports.ComponentDefinition = interfaces_1.ComponentDefinition;
exports.ComponentLayoutBuilder = interfaces_1.ComponentLayoutBuilder;
exports.ComponentAttrsBuilder = interfaces_1.ComponentAttrsBuilder;
exports.isComponentDefinition = interfaces_1.isComponentDefinition;
var interfaces_2 = require("./lib/modifier/interfaces");
exports.ModifierManager = interfaces_2.ModifierManager;
var helper_1 = require("./lib/dom/helper");
exports.DOMChanges = helper_1.default;
exports.IDOMChanges = helper_1.DOMChanges;
exports.DOMTreeConstruction = helper_1.DOMTreeConstruction;
exports.isWhitespace = helper_1.isWhitespace;
exports.insertHTMLBefore = helper_1.insertHTMLBefore;
var Simple = require("./lib/dom/interfaces");
exports.Simple = Simple;
var builder_2 = require("./lib/builder");
exports.ElementStack = builder_2.ElementStack;
exports.ElementOperations = builder_2.ElementOperations;
var bounds_1 = require("./lib/bounds");
exports.Bounds = bounds_1.default;
exports.ConcreteBounds = bounds_1.ConcreteBounds;
