// @glimmer/object-reference/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var descriptors_1 = require("./lib/references/descriptors");
exports.ComputedReferenceBlueprint = descriptors_1.ComputedReferenceBlueprint;
exports.InnerReferenceFactory = descriptors_1.InnerReferenceFactory;
exports.PropertyReference = descriptors_1.PropertyReference;
var meta_1 = require("./lib/meta");
exports.CLASS_META = meta_1.CLASS_META;
exports.Meta = meta_1.default;
exports.metaFor = meta_1.metaFor;
var object_1 = require("./lib/object");
exports.setProperty = object_1.setProperty;
exports.notifyProperty = object_1.notifyProperty;
__export(require("./lib/types"));
var path_1 = require("./lib/references/path");
exports.ObjectReference = path_1.default;
var root_1 = require("./lib/references/root");
exports.UpdatableReference = root_1.default;
var reference_1 = require("@glimmer/reference");
exports.ConstReference = reference_1.ConstReference;
exports.isConst = reference_1.isConst;
