// glimmer-engine/packages/glimmer-object-model/dist/modules/lib/reference.js
import Meta, { ClassMeta } from './meta';
import { CURRENT_TAG, CONSTANT_TAG, combine } from 'glimmer-reference';
const META = new WeakMap();
const CLASS_META = new WeakMap();
export function classMeta(object) {
    let m = CLASS_META.get(object);
    if (m === undefined) {
        m = new ClassMeta();
        CLASS_META.set(object, m);
    }
    return m;
}
export function meta(object) {
    let m = META.get(object);
    if (m === undefined) {
        m = new Meta();
        META.set(object, m);
    }
    return m;
}
export function set(object, key, value) {
    object[key] = value;
    meta(object).dirty(key);
}
export function root(object) {
    return new VersionedRootReference(object);
}
export class VersionedRootReference {
    constructor(inner) {
        this.inner = inner;
        this.tag = CONSTANT_TAG;
    }
    value() {
        return this.inner;
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
export class VersionedObjectReference {
    constructor(parent, key) {
        this.parent = parent;
        this.key = key;
        this.tag = CURRENT_TAG;
    }
    value() {
        let { parent, key } = this;
        let parentObject = this.parent.value();
        let computed = classMeta(Object.getPrototypeOf(parentObject)).getComputed(key);
        let tags = [meta(parentObject).tag(key)];
        if (computed) {
            tags.push(...computed.dependentKeys.map(key => path(this, key).tag));
        }
        this.tag = combine(tags);
        return parentObject[key];
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
function path(parent, key) {
    return key.split('.').reduce((ref, part) => ref.get(part), parent);
}
