// glimmer-engine/dist/named-amd/@glimmer/util/index.js
enifed('@glimmer/util/index', ['exports', '@glimmer/util/lib/namespaces', '@glimmer/util/lib/platform-utils', '@glimmer/util/lib/assert', '@glimmer/util/lib/logger', '@glimmer/util/lib/object-utils', '@glimmer/util/lib/guid', '@glimmer/util/lib/collections', '@glimmer/util/lib/list-utils', '@glimmer/util/lib/array-utils', '@glimmer/util/lib/weakmap'], function (exports, _glimmerUtilLibNamespaces, _glimmerUtilLibPlatformUtils, _glimmerUtilLibAssert, _glimmerUtilLibLogger, _glimmerUtilLibObjectUtils, _glimmerUtilLibGuid, _glimmerUtilLibCollections, _glimmerUtilLibListUtils, _glimmerUtilLibArrayUtils, _glimmerUtilLibWeakmap) {
  'use strict';

  function _interopExportWildcard(obj, defaults) { var newObj = defaults({}, obj); delete newObj['default']; return newObj; }

  function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

  exports.getAttrNamespace = _glimmerUtilLibNamespaces.getAttrNamespace;

  _defaults(exports, _interopExportWildcard(_glimmerUtilLibPlatformUtils, _defaults));

  exports.assert = _glimmerUtilLibAssert.default;
  exports.LOGGER = _glimmerUtilLibLogger.default;
  exports.Logger = _glimmerUtilLibLogger.Logger;
  exports.LogLevel = _glimmerUtilLibLogger.LogLevel;
  exports.assign = _glimmerUtilLibObjectUtils.assign;
  exports.fillNulls = _glimmerUtilLibObjectUtils.fillNulls;
  exports.ensureGuid = _glimmerUtilLibGuid.ensureGuid;
  exports.initializeGuid = _glimmerUtilLibGuid.initializeGuid;
  exports.HasGuid = _glimmerUtilLibGuid.HasGuid;
  exports.Stack = _glimmerUtilLibCollections.Stack;
  exports.Dict = _glimmerUtilLibCollections.Dict;
  exports.Set = _glimmerUtilLibCollections.Set;
  exports.DictSet = _glimmerUtilLibCollections.DictSet;
  exports.dict = _glimmerUtilLibCollections.dict;
  exports.EMPTY_SLICE = _glimmerUtilLibListUtils.EMPTY_SLICE;
  exports.LinkedList = _glimmerUtilLibListUtils.LinkedList;
  exports.LinkedListNode = _glimmerUtilLibListUtils.LinkedListNode;
  exports.ListNode = _glimmerUtilLibListUtils.ListNode;
  exports.CloneableListNode = _glimmerUtilLibListUtils.CloneableListNode;
  exports.ListSlice = _glimmerUtilLibListUtils.ListSlice;
  exports.Slice = _glimmerUtilLibListUtils.Slice;
  exports.A = _glimmerUtilLibArrayUtils.default;
  exports.HAS_NATIVE_WEAKMAP = _glimmerUtilLibWeakmap.HAS_NATIVE_WEAKMAP;
});
