// glimmer-engine/packages/glimmer-object-model/dist/commonjs/lib/reference.js
"use strict";
const meta_1 = require("./meta");
const glimmer_reference_1 = require("glimmer-reference");
const META = new WeakMap();
const CLASS_META = new WeakMap();
function classMeta(object) {
    let m = CLASS_META.get(object);
    if (m === undefined) {
        m = new meta_1.ClassMeta();
        CLASS_META.set(object, m);
    }
    return m;
}
exports.classMeta = classMeta;
function meta(object) {
    let m = META.get(object);
    if (m === undefined) {
        m = new meta_1.default();
        META.set(object, m);
    }
    return m;
}
exports.meta = meta;
function set(object, key, value) {
    object[key] = value;
    meta(object).dirty(key);
}
exports.set = set;
function root(object) {
    return new VersionedRootReference(object);
}
exports.root = root;
class VersionedRootReference {
    constructor(inner) {
        this.inner = inner;
        this.tag = glimmer_reference_1.CONSTANT_TAG;
    }
    value() {
        return this.inner;
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
exports.VersionedRootReference = VersionedRootReference;
class VersionedObjectReference {
    constructor(parent, key) {
        this.parent = parent;
        this.key = key;
        this.tag = glimmer_reference_1.CURRENT_TAG;
    }
    value() {
        let { parent, key } = this;
        let parentObject = this.parent.value();
        let computed = classMeta(Object.getPrototypeOf(parentObject)).getComputed(key);
        let tags = [meta(parentObject).tag(key)];
        if (computed) {
            tags.push(...computed.dependentKeys.map(key => path(this, key).tag));
        }
        this.tag = glimmer_reference_1.combine(tags);
        return parentObject[key];
    }
    get(key) {
        return new VersionedObjectReference(this, key);
    }
}
exports.VersionedObjectReference = VersionedObjectReference;
function path(parent, key) {
    return key.split('.').reduce((ref, part) => ref.get(part), parent);
}
