// glimmer-engine/packages/@glimmer/object-model/dist/modules/lib/blueprint.js
import { classMeta } from './reference';
export class Blueprint {
    constructor(key) {
        this.key = key;
    }
}
const VALUE_DESCRIPTOR = {
    enumerable: true,
    configurable: true,
    writable: true,
    value: undefined
};
const ACCESSOR_DESCRIPTOR = {
    enumerable: true,
    configurable: true,
    get: undefined,
    set: undefined
};
export class ValueBlueprint extends Blueprint {
    constructor(key, value) {
        super(key);
        this.value = value;
    }
    define(home) {
        VALUE_DESCRIPTOR.value = this.value;
        Object.defineProperty(home, this.key, VALUE_DESCRIPTOR);
    }
}
export class GlimmerDescriptor {
}
export class Computed extends GlimmerDescriptor {
    constructor(dependentKeys, accessor) {
        super();
        this.dependentKeys = dependentKeys;
        this.accessor = accessor;
    }
    blueprint(key) {
        return new ComputedBlueprint(key, this);
    }
    reference(root, key) {
    }
    define(home, key) {
        classMeta(home).defineComputed(key, this);
        ACCESSOR_DESCRIPTOR.get = this.accessor.get;
        ACCESSOR_DESCRIPTOR.set = this.accessor.set;
        Object.defineProperty(home, key, ACCESSOR_DESCRIPTOR);
    }
}
export function computed(...args) {
    let depKeys = args.slice(0, -1);
    let accessor = args[args.length - 1];
    return new Computed(depKeys, accessor);
}
export class ComputedBlueprint extends Blueprint {
    constructor(key, computed) {
        super(key);
        this.computed = computed;
    }
    define(home) {
        this.computed.define(home, this.key);
    }
}
export class DescriptorBlueprint extends Blueprint {
    constructor(key, descriptor) {
        super(key);
        this.descriptor = descriptor;
    }
    define(home) {
        Object.defineProperty(home, this.key, this.descriptor);
    }
}
